{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1efe622f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import hamming\n",
    "import gudhi as gd\n",
    "import seaborn as sns\n",
    "from statistics import mode\n",
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "import gudhi as gd\n",
    "from scipy.spatial.distance import hamming\n",
    "import plotly.graph_objs as go\n",
    "import networkx as nx\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "from networkx.utils import not_implemented_for, pairwise\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87686cb0-17ca-4bb1-998b-a3037402531f",
   "metadata": {},
   "source": [
    "<h3>Dos funciones para obtener los data frames de presencia</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c59d1c1a-8fde-4959-89bf-2bd46357d881",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_info_by_pathway_and_all_genomes_directions(info):\n",
    "    data_frames={}\n",
    "    pathways=set()\n",
    "    directions={}\n",
    "    genome_directions={}\n",
    "    for i in range(info.shape[0]):\n",
    "        #Revisar si se debe separar esto\n",
    "        #pathway=info.loc[i,0].split('|')[0].strip('\\ufeff')\n",
    "        pathway=info.loc[i,0].split('|')[0]\n",
    "        if pathway not in directions.keys():\n",
    "            directions[pathway]=[i]\n",
    "        else:\n",
    "            directions[pathway].append(i)\n",
    "        #Se cambia la primer columna, para tener los genes\n",
    "        gen_with_number=info.loc[i,0].split('|')[2]\n",
    "        gen=gen_with_number[0:gen_with_number.rfind('_')]\n",
    "        info.loc[i,0]=gen\n",
    "        #Se obtienen los genomas\n",
    "        genome=info.loc[i,1].split('|')[2]\n",
    "        if genome not in genome_directions.keys():\n",
    "            genome_directions[genome]=[i]\n",
    "        else:\n",
    "            genome_directions[genome].append(i)\n",
    "\n",
    "    \n",
    "    for pathway in directions.keys():\n",
    "        data_frames[pathway]=info.loc[directions[pathway],:]\n",
    "    return data_frames,genome_directions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "4136747a-2e6c-4143-97a9-0beb91adcbf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_by_pathway(info_by_pathway_and_genomes_directions):\n",
    "    info_by_pathway=info_by_pathway_and_genomes_directions[0]\n",
    "    genomes_directions=info_by_pathway_and_genomes_directions[1]\n",
    "    df_by_pathway={}\n",
    "\n",
    "    for path in info_by_pathway.keys():\n",
    "        info_path=info_by_pathway[path]\n",
    "        genes=set(info_by_pathway[path][0].values)\n",
    "        df=pd.DataFrame(index=genomes_directions.keys(),columns=list(genes))\n",
    "        df.index.name=path\n",
    "        \n",
    "        for gen in genes:\n",
    "            info_path_and_gen=info_path[info_path[0]==gen]\n",
    "            for genom in genomes_directions.keys():\n",
    "                direction=set(genomes_directions[genom]) & set(info_path_and_gen.index) \n",
    "                genomes_per_gen=info_path_and_gen.loc[list(direction)]\n",
    "                uniq_genomes_per_gen=genomes_per_gen[1].drop_duplicates()\n",
    "                df.loc[genom,gen]=uniq_genomes_per_gen.shape[0]\n",
    "        df_by_pathway[path]=df.sort_index()\n",
    "    return df_by_pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07b9660-946d-445f-967d-df13f5da2fc0",
   "metadata": {},
   "source": [
    "<h3>Funciones para dar peso</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1620fc4-199b-415f-97dd-d4a86539dc97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mode(df):\n",
    "    arr=[]\n",
    "    for i in df.values:\n",
    "        for j in i:\n",
    "            if j !=0:\n",
    "                arr.append(j)\n",
    "    return mode(arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1b51b1e2-10b8-40ca-89a6-a6c09c4e5e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_weights(presence):\n",
    "    mode=get_mode(presence)\n",
    "    #peso menor que moda\n",
    "    piv=presence[presence<mode]\n",
    "    presence[piv>0]=1\n",
    "    #peso igual a moda\n",
    "    presence[presence==mode]=2\n",
    "    #peso mayor que moda\n",
    "    presence[presence>mode]=3\n",
    "    return presence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a8a4950-187e-4d1b-b3c5-16ed8e62fa1b",
   "metadata": {},
   "source": [
    "<h2>Funciones para crear matrices de distancia,los complejos simpliciales</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4c7d06f3-9959-4a06-8651-304a7a075479",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's assume that \"population\" is a numpy ndarray with your genomes as rows.\n",
    "def calculate_hamming_matrix(population):\n",
    "    # Number of genomes\n",
    "    num_genomes = population.shape[0]\n",
    "    # Create an empty matrix for Hamming distances\n",
    "    hamming_matrix = np.zeros((num_genomes, num_genomes), dtype=int)\n",
    "   # Calculate the Hamming distance between each pair of genomes\n",
    "    for i in range(num_genomes):\n",
    "        for j in range(i+1, num_genomes):  # j=i+1 to avoid calculating the same distance twice\n",
    "            # The Hamming distance is multiplied by the number of genes to convert it into an absolute distance\n",
    "            distance = hamming(population[i], population[j]) * len(population[i])\n",
    "            hamming_matrix[i, j] = distance\n",
    "            hamming_matrix[j, i] = distance  # The matrix is symmetric\n",
    "    \n",
    "    return hamming_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "14a245e5-bfd1-4377-9857-a9531abed9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_complex(distance_matrix2):\n",
    "    # Create the Rips simplicial complex from the distance matrix\n",
    "    rips_complex = gd.RipsComplex(distance_matrix=distance_matrix2)\n",
    "    # Create the simplex tree from the Rips complex with a maximum dimension of 3\n",
    "    simplex_tree = rips_complex.create_simplex_tree(max_dimension=2)\n",
    "    # Compute the persistence of the simplicial complex\n",
    "    persistence = simplex_tree.persistence()\n",
    "    # Return the persistence diagram or barcode\n",
    "    return persistence, simplex_tree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ec9bd8f1-de8b-472f-ba7a-a2d6ff870d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_complex_by_pathways(path_dict):\n",
    "    complex_path_dict={}\n",
    "    for path in path_dict.keys():\n",
    "        complex_path_dict[path]=create_complex(calculate_hamming_matrix(path_dict[path].values))\n",
    "    return complex_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "39835cc1-fd33-4c54-9990-43a5378396bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_all_bar_code_pathways(complex_pathways):\n",
    "    for i in complex_pathways.keys():\n",
    "        gd.plot_persistence_barcode(complex_pathways[i][0],legend=True)\n",
    "        plt.title(f\"{i}\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "795532de-e8cf-46d2-a1ca-3368d4ca5003",
   "metadata": {},
   "source": [
    "<h2>Funciones para obtener los uno-hoyos</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "fa6ef6c2-412c-4799-bfea-594f46f83414",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dict_holes_to_names(holes):\n",
    "    key=list(df_by_pathway.keys())[0]\n",
    "    df=df_by_pathway[key].copy(deep=True)\n",
    "    df.reset_index(inplace=True)\n",
    "    true_name_holes=[]\n",
    "    for hole in holes:\n",
    "        true_name_holes.append(list(df.loc[hole,:]['index']))\n",
    "    return true_name_holes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d8cabcaf-d029-433e-81e3-69de8ef098a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimum_cycle_basis(G, weight=None, total=None):\n",
    "    \"\"\"Returns a minimum weight cycle basis for G\n",
    "\n",
    "    Minimum weight means a cycle basis for which the total weight\n",
    "    (length for unweighted graphs) of all the cycles is minimum.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    G : NetworkX Graph\n",
    "    weight: string\n",
    "        name of the edge attribute to use for edge weights\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    A list of cycle lists.  Each cycle list is a list of nodes\n",
    "    which forms a cycle (loop) in G. Note that the nodes are not\n",
    "    necessarily returned in a order by which they appear in the cycle\n",
    "\n",
    "    Examples\n",
    "    --------\n",
    "    >>> G = nx.Graph()\n",
    "    >>> nx.add_cycle(G, [0, 1, 2, 3])\n",
    "    >>> nx.add_cycle(G, [0, 3, 4, 5])\n",
    "    >>> nx.minimum_cycle_basis(G)\n",
    "    [[5, 4, 3, 0], [3, 2, 1, 0]]\n",
    "\n",
    "    References:\n",
    "        [1] Kavitha, Telikepalli, et al. \"An O(m^2n) Algorithm for\n",
    "        Minimum Cycle Basis of Graphs.\"\n",
    "        http://link.springer.com/article/10.1007/s00453-007-9064-z\n",
    "        [2] de Pina, J. 1995. Applications of shortest path methods.\n",
    "        Ph.D. thesis, University of Amsterdam, Netherlands\n",
    "\n",
    "    See Also\n",
    "    --------\n",
    "    simple_cycles, cycle_basis\n",
    "    \"\"\"\n",
    "    # We first split the graph in connected subgraphs\n",
    "    return sum(\n",
    "        (_min_cycle_basis(G.subgraph(c), weight,total) for c in nx.connected_components(G)),\n",
    "        [],\n",
    "    )\n",
    "\n",
    "def _min_cycle_basis(G, weight,total):\n",
    "    cb = []\n",
    "    cont=0\n",
    "    # We  extract the edges not in a spanning tree. We do not really need a\n",
    "    # *minimum* spanning tree. That is why we call the next function with\n",
    "    # weight=None. Depending on implementation, it may be faster as well\n",
    "    tree_edges = list(nx.minimum_spanning_edges(G, weight=None, data=False))\n",
    "    chords = G.edges - tree_edges - {(v, u) for u, v in tree_edges}\n",
    "\n",
    "    # We maintain a set of vectors orthogonal to sofar found cycles\n",
    "    set_orth = [{edge} for edge in chords]\n",
    "    while set_orth:\n",
    "        if cont==total:\n",
    "            break\n",
    "        base = set_orth.pop()\n",
    "        # kth cycle is \"parallel\" to kth vector in set_orth\n",
    "        cycle_edges = _min_cycle(G, base, weight)\n",
    "        cycle_found=[v for u, v in cycle_edges]\n",
    "\n",
    "        if len(cycle_found)>3:\n",
    "            cb.append(cycle_found)\n",
    "            cont+=1\n",
    "\n",
    "        # now update set_orth so that k+1,k+2... th elements are\n",
    "        # orthogonal to the newly found cycle, as per [p. 336, 1]\n",
    "        set_orth = [\n",
    "            (\n",
    "                {e for e in orth if e not in base if e[::-1] not in base}\n",
    "                | {e for e in base if e not in orth if e[::-1] not in orth}\n",
    "            )\n",
    "            if sum((e in orth or e[::-1] in orth) for e in cycle_edges) % 2\n",
    "            else orth\n",
    "            for orth in set_orth\n",
    "        ]\n",
    "    return cb\n",
    "\n",
    "\n",
    "def _min_cycle(G, orth, weight):\n",
    "    \"\"\"\n",
    "    Computes the minimum weight cycle in G,\n",
    "    orthogonal to the vector orth as per [p. 338, 1]\n",
    "    Use (u, 1) to indicate the lifted copy of u (denoted u' in paper).\n",
    "    \"\"\"\n",
    "    Gi = nx.Graph()\n",
    "\n",
    "    # Add 2 copies of each edge in G to Gi.\n",
    "    # If edge is in orth, add cross edge; otherwise in-plane edge\n",
    "    for u, v, wt in G.edges(data=weight, default=1):\n",
    "        if (u, v) in orth or (v, u) in orth:\n",
    "            Gi.add_edges_from([(u, (v, 1)), ((u, 1), v)], Gi_weight=wt)\n",
    "        else:\n",
    "            Gi.add_edges_from([(u, v), ((u, 1), (v, 1))], Gi_weight=wt)\n",
    "\n",
    "    # find the shortest length in Gi between n and (n, 1) for each n\n",
    "    # Note: Use \"Gi_weight\" for name of weight attribute\n",
    "    spl = nx.shortest_path_length\n",
    "    lift = {n: spl(Gi, source=n, target=(n, 1), weight=\"Gi_weight\") for n in G}\n",
    "\n",
    "    # Now compute that short path in Gi, which translates to a cycle in G\n",
    "    start = min(lift, key=lift.get)\n",
    "    end = (start, 1)\n",
    "    min_path_i = nx.shortest_path(Gi, source=start, target=end, weight=\"Gi_weight\")\n",
    "\n",
    "    # Now we obtain the actual path, re-map nodes in Gi to those in G\n",
    "    min_path = [n if n in G else n[0] for n in min_path_i]\n",
    "\n",
    "    # Now remove the edges that occur two times\n",
    "    # two passes: flag which edges get kept, then build it\n",
    "    edgelist = list(pairwise(min_path))\n",
    "    edgeset = set()\n",
    "    for e in edgelist:\n",
    "        if e in edgeset:\n",
    "            edgeset.remove(e)\n",
    "        elif e[::-1] in edgeset:\n",
    "            edgeset.remove(e[::-1])\n",
    "        else:\n",
    "            edgeset.add(e)\n",
    "\n",
    "    min_edgelist = []\n",
    "    for e in edgelist:\n",
    "        if e in edgeset:\n",
    "            min_edgelist.append(e)\n",
    "            edgeset.remove(e)\n",
    "        elif e[::-1] in edgeset:\n",
    "            min_edgelist.append(e[::-1])\n",
    "            edgeset.remove(e[::-1])\n",
    "\n",
    "    return min_edgelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ca0600d2-b1d6-4364-891b-7cb2ff739653",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_all_cycles(complex_path):\n",
    "    persistence,simplex_tree=complex_pathways[complex_path]\n",
    "    born_and_number = born_filtraton_value_holes(persistence)\n",
    "    G=nx.Graph()\n",
    "    born=born_and_number.keys()\n",
    "    ciclos_dep=set()\n",
    "    filtration=0\n",
    "    for simplex, filt in simplex_tree.get_filtration():\n",
    "        #if len(ciclos_dep)==num_holes:\n",
    "         #   break\n",
    "        \n",
    "        if filtration!=filt and filtration in born:\n",
    "            number=born_and_number[filtration]\n",
    "            print('se buscan ciclos en el tiempo', filt,'para',complex_path)\n",
    "            ciclos=minimum_cycle_basis(G,total=number)\n",
    "            for ciclo in ciclos:\n",
    "                if len(ciclo)>3:\n",
    "                    print('Se encontró el ciclo',ciclo,'en el tiempo', filtration,'para',complex_path)\n",
    "                    ciclos_dep.add(tuple(ciclo))\n",
    "                    #Se llena el hoyo\n",
    "                    for i in ciclo:\n",
    "                        for j in ciclo:\n",
    "                            G.add_edge(i,j)\n",
    "                            \n",
    "            \n",
    "        filtration=filt\n",
    "        \n",
    "        if len(simplex)==2:\n",
    "            G.add_edge(simplex[0], simplex[1])\n",
    "\n",
    "\n",
    "    #direc='pseudomonas_'+complex_path+'.txt'\n",
    "    #f = open(direc, \"a\")\n",
    "    #f.write(str(ciclos_dep))\n",
    "    #f.close()\n",
    "        \n",
    "\n",
    "    return [complex_path,dict_holes_to_names(ciclos_dep)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c7ac9e02-f05e-4fae-aaa5-e2203d70ba10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def born_filtraton_value_holes(persistence):\n",
    "    born=[]\n",
    "    for bar in persistence:\n",
    "        if bar[0]==1:\n",
    "            born.append(bar[1][0])\n",
    "            \n",
    "    born_and_number=set([(x,born.count(x)) for x  in born])\n",
    "\n",
    "    return dict(born_and_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1fae47c4-0658-4ef8-89a3-752fdb69612a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_holes_by_pathways(complex_by_path):\n",
    "    llaves=complex_by_path.keys()\n",
    "    with multiprocessing.Pool() as pool:\n",
    "        # Mapear los ítems a la función process_item en paralelo\n",
    "        results = pool.map(find_all_cycles,llaves)\n",
    "    return dict(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58a76d6-d678-4259-bc55-80638f1b1eb9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cb97e833-ec03-4170-bdd1-aa1160028e4f",
   "metadata": {},
   "source": [
    "<h2>EJECUCIÓN:Se pide un score minimo</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "cee256e2-988f-44ee-b1da-1a18dc400aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "info=pd.read_csv('data/pscpPseudo.blast',sep='\\t',header=None)\n",
    "minimum_score=0\n",
    "info=info[info[11]>=minimum_score]\n",
    "info=info.loc[:,0:1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c51425f8-9298-4f16-869f-eda6dd6f6b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_by_pathway_and_genomes_directions=get_info_by_pathway_and_all_genomes_directions(info)\n",
    "df_by_pathway=get_df_by_pathway(info_by_pathway_and_genomes_directions)\n",
    "complex_pathways=get_complex_by_pathways(df_by_pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "944fcf8b-f628-4712-b7e2-38bd46941ae8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "se buscan ciclos en el tiempo 2.0 parase buscan ciclos en el tiempo se buscan ciclos en el tiempoALPHAKETOGLUTARATE_AMINOACIDS \n",
      "2.0  2.0para  paraPYR_THR_AA \n",
      "OXALACETATE_AMINOACIDS\n",
      "se buscan ciclos en el tiempo 2.0 para 3PGA_AMINOACIDS\n",
      "se buscan ciclos en el tiempo 2.0 para E4P_AMINO_ACIDS\n",
      "se buscan ciclos en el tiempo 2.0 para R5P_AMINOACIDS\n",
      "se buscan ciclos en el tiempo 2.0 para TCA\n",
      "se buscan ciclos en el tiempo 2.0 para Glycolysis\n",
      "Se encontró el ciclo [111, 120, 79, 131] en el tiempo 1.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "Se encontró el ciclo [70, 102, 69, 7] en el tiempo 1.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "Se encontró el ciclo [69, 105, 24, 137] en el tiempo 1.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "Se encontró el ciclo [82, 49, 71, 7] en el tiempo 1.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "Se encontró el ciclo [82, 145, 70, 7] en el tiempo 1.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "se buscan ciclos en el tiempo 3.0 para ALPHAKETOGLUTARATE_AMINOACIDS\n",
      "Se encontró el ciclo [105, 34, 23, 135] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [122, 13, 30, 26] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [41, 123, 11, 26] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [5, 197, 0, 12] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [120, 130, 31, 14] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [43, 46, 39, 0] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [129, 17, 42, 5, 12, 0, 39, 14] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [166, 39, 46, 33] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [176, 154, 122, 26] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [177, 8, 13, 30] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [120, 58, 94, 8, 13, 166, 39, 14] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [30, 177, 38, 96, 11, 26] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [68, 109, 0, 12] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [177, 30, 41, 25] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [129, 58, 120, 14] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [47, 114, 18, 37] en el tiempo 1.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [94, 58, 17, 42] en el tiempo 1.0 para PYR_THR_AA\n",
      "se buscan ciclos en el tiempo 3.0 para PYR_THR_AA\n",
      "Se encontró el ciclo [119, 105, 117, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [177, 51, 119, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [176, 47, 128, 71, 25, 90, 55, 41] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [52, 37, 31, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [88, 108, 26, 36] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [74, 50, 36, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [119, 108, 88, 51] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [155, 3, 146, 197, 30, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [38, 20, 25, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [177, 90, 25, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [164, 20, 38, 108, 26, 30] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [166, 11, 96, 58, 38, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [141, 142, 91, 160, 75, 117] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [55, 74, 18, 41] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [195, 100, 186, 10] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [36, 26, 30, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [209, 20, 38, 9] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [119, 108, 38, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [90, 142, 91, 31, 74, 55] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [74, 18, 52, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [146, 197, 164, 20, 25, 85, 60, 168, 7] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [187, 186, 57, 184] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [90, 25, 20, 164, 74, 55] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [31, 91, 0, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [36, 82, 31, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [168, 60, 69, 112] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [203, 198, 195, 10] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [74, 164, 30, 1] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [135, 60, 85, 117] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [146, 62, 151, 59, 28, 167, 45, 202, 60, 168, 7] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [85, 60, 168, 49, 51, 119, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [141, 69, 135, 117] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [177, 142, 141, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "Se encontró el ciclo [117, 75, 38, 71] en el tiempo 1.0 para OXALACETATE_AMINOACIDS\n",
      "se buscan ciclos en el tiempo 3.0 para OXALACETATE_AMINOACIDS\n"
     ]
    }
   ],
   "source": [
    "holes=get_holes_by_pathways(complex_by_pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "c99245c4-934a-4914-87a0-8ed304b59fdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['666666.100019', '666666.100028', '666666.100027', '666666.100006'],\n",
       " ['666666.100023', '666666.100031', '666666.100007', '666666.100004'],\n",
       " ['666666.100023', '666666.100030', '666666.100006', '666666.100004']]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "holes['Glycolysis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "370a276a-bd95-4765-9874-165406ef7b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_by_pathway['TCA'].to_csv('TCA.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "1af71602-167b-44fb-a8fc-9aa78877efcf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>glyceraldehyde3phosphate_dehydrogenase</th>\n",
       "      <th>Fructose_biphosphate_aldolase</th>\n",
       "      <th>Glucose_kinase</th>\n",
       "      <th>Triosephosphate_isomerase</th>\n",
       "      <th>phosphoglycerate_kinase</th>\n",
       "      <th>phosphoglycerate_mutase</th>\n",
       "      <th>Phosphofructokinase</th>\n",
       "      <th>Glucose_6_phosphate_isomerase</th>\n",
       "      <th>Pyruvate_kinase</th>\n",
       "      <th>enolase</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>666666.100004</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100005</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100006</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100007</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100008</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100010</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100019</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100020</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100023</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100025</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100026</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100027</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100028</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100029</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100030</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100031</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100032</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100040</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100041</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100042</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666666.100044</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              glyceraldehyde3phosphate_dehydrogenase  \\\n",
       "G                                                      \n",
       "666666.100004                                      1   \n",
       "666666.100005                                      1   \n",
       "666666.100006                                      1   \n",
       "666666.100007                                      1   \n",
       "666666.100008                                      1   \n",
       "666666.100010                                      1   \n",
       "666666.100019                                      1   \n",
       "666666.100020                                      1   \n",
       "666666.100023                                      1   \n",
       "666666.100025                                      1   \n",
       "666666.100026                                      1   \n",
       "666666.100027                                      1   \n",
       "666666.100028                                      1   \n",
       "666666.100029                                      1   \n",
       "666666.100030                                      1   \n",
       "666666.100031                                      1   \n",
       "666666.100032                                      1   \n",
       "666666.100040                                      1   \n",
       "666666.100041                                      1   \n",
       "666666.100042                                      0   \n",
       "666666.100044                                      1   \n",
       "\n",
       "              Fructose_biphosphate_aldolase Glucose_kinase  \\\n",
       "G                                                            \n",
       "666666.100004                             1              4   \n",
       "666666.100005                             1              5   \n",
       "666666.100006                             1              4   \n",
       "666666.100007                             1              4   \n",
       "666666.100008                             1              3   \n",
       "666666.100010                             1              4   \n",
       "666666.100019                             1              6   \n",
       "666666.100020                             1              8   \n",
       "666666.100023                             1              6   \n",
       "666666.100025                             1              2   \n",
       "666666.100026                             1              7   \n",
       "666666.100027                             1              4   \n",
       "666666.100028                             1              3   \n",
       "666666.100029                             0              1   \n",
       "666666.100030                             1              7   \n",
       "666666.100031                             1              5   \n",
       "666666.100032                             1              7   \n",
       "666666.100040                             2              5   \n",
       "666666.100041                             2              3   \n",
       "666666.100042                             1              3   \n",
       "666666.100044                             1              6   \n",
       "\n",
       "              Triosephosphate_isomerase phosphoglycerate_kinase  \\\n",
       "G                                                                 \n",
       "666666.100004                         1                       2   \n",
       "666666.100005                         1                       2   \n",
       "666666.100006                         1                       2   \n",
       "666666.100007                         1                       2   \n",
       "666666.100008                         1                       2   \n",
       "666666.100010                         1                       2   \n",
       "666666.100019                         1                       2   \n",
       "666666.100020                         1                       2   \n",
       "666666.100023                         1                       2   \n",
       "666666.100025                         2                       2   \n",
       "666666.100026                         1                       2   \n",
       "666666.100027                         1                       2   \n",
       "666666.100028                         1                       2   \n",
       "666666.100029                         1                       2   \n",
       "666666.100030                         1                       2   \n",
       "666666.100031                         1                       2   \n",
       "666666.100032                         1                       2   \n",
       "666666.100040                         1                       2   \n",
       "666666.100041                         1                       2   \n",
       "666666.100042                         0                       1   \n",
       "666666.100044                         1                       2   \n",
       "\n",
       "              phosphoglycerate_mutase Phosphofructokinase  \\\n",
       "G                                                           \n",
       "666666.100004                       5                   1   \n",
       "666666.100005                       4                   1   \n",
       "666666.100006                       3                   1   \n",
       "666666.100007                       4                   1   \n",
       "666666.100008                       5                   1   \n",
       "666666.100010                       4                   2   \n",
       "666666.100019                       3                   2   \n",
       "666666.100020                       4                   2   \n",
       "666666.100023                       5                   3   \n",
       "666666.100025                       6                   2   \n",
       "666666.100026                       6                   3   \n",
       "666666.100027                       3                   1   \n",
       "666666.100028                       3                   2   \n",
       "666666.100029                       2                   0   \n",
       "666666.100030                       3                   3   \n",
       "666666.100031                       4                   3   \n",
       "666666.100032                       4                   3   \n",
       "666666.100040                       3                   2   \n",
       "666666.100041                       2                   3   \n",
       "666666.100042                       3                   1   \n",
       "666666.100044                       3                   3   \n",
       "\n",
       "              Glucose_6_phosphate_isomerase Pyruvate_kinase enolase  \n",
       "G                                                                    \n",
       "666666.100004                             0               1       1  \n",
       "666666.100005                             0               1       1  \n",
       "666666.100006                             0               1       1  \n",
       "666666.100007                             0               1       1  \n",
       "666666.100008                             0               1       1  \n",
       "666666.100010                             0               1       1  \n",
       "666666.100019                             0               1       1  \n",
       "666666.100020                             0               1       1  \n",
       "666666.100023                             0               1       1  \n",
       "666666.100025                             0               1       1  \n",
       "666666.100026                             0               1       1  \n",
       "666666.100027                             1               1       1  \n",
       "666666.100028                             1               1       1  \n",
       "666666.100029                             0               0       0  \n",
       "666666.100030                             0               1       1  \n",
       "666666.100031                             0               1       1  \n",
       "666666.100032                             0               1       1  \n",
       "666666.100040                             0               1       1  \n",
       "666666.100041                             1               1       1  \n",
       "666666.100042                             0               0       0  \n",
       "666666.100044                             0               1       1  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_by_pathway['Glycolysis'].index.name='G'\n",
    "df_by_pathway['Glycolysis']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TDA",
   "language": "python",
   "name": "tda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
