{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1efe622f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import hamming\n",
    "import gudhi as gd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84be2c16",
   "metadata": {},
   "source": [
    "<h3>Se cargan los datos</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d451b8ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Se cargan las coincidencias obtenidas por el BBH\n",
    "pseudo_csv=pd.read_csv('pseudo.csv', sep='\\t', header=None,index_col=0)\n",
    "\n",
    "#Se cargan los nombres comunes de las especies\n",
    "#pseudo_nombres=pd.read_csv('/files/francisco_esp/Pseudomonas/Pseudomonas.ids', sep='\\t', header=None, dtype=str)\n",
    "#pseudo_nombres.set_index(1,inplace=True)\n",
    "pseudo_csv.drop(np.NaN,inplace=True)\n",
    "pseudo_csv.reset_index(inplace=True)\n",
    "pseudo_csv.drop(columns=0,inplace=True)\n",
    "pseudo_csv\n",
    "\n",
    "pseudo_info=pseudo_csv[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c66eb336-851d-4277-aecd-8336ceadf56d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6666666.136315.1646 -&gt; calcium-dependent-antib...</td>\n",
       "      <td>gi|6666666.136315.1646|6666666.136315|AHZY01|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6666666.134953.3589 -&gt; polyketide|215|215_1|21...</td>\n",
       "      <td>gi|6666666.134953.3589|6666666.134953|AKCL01|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6666666.136574.4344 -&gt; OXALACETATE_AMINOACIDS|...</td>\n",
       "      <td>gi|6666666.136574.4344|6666666.136574|ATKI01|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6666666.135439.1862 -&gt; 3PGA_AMINOACIDS|6|Cyste...</td>\n",
       "      <td>gi|6666666.135439.1862|6666666.135439|JXDG01|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6666666.136502.4654 -&gt; ALPHAKETOGLUTARATE_AMIN...</td>\n",
       "      <td>gi|6666666.136502.4654|6666666.136502|ATLO01|A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12128</th>\n",
       "      <td>6666666.102116.464 -&gt; ALPHAKETOGLUTARATE_AMINO...</td>\n",
       "      <td>gi|6666666.102116.464|6666666.102116|NC_015379...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12129</th>\n",
       "      <td>6666666.94971.5998 -&gt; ALPHAKETOGLUTARATE_AMINO...</td>\n",
       "      <td>gi|6666666.94971.5998|6666666.94971|201|Glutam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12130</th>\n",
       "      <td>6666666.102121.2566 -&gt; ALPHAKETOGLUTARATE_AMIN...</td>\n",
       "      <td>gi|6666666.102121.2566|6666666.102121|AJXG01|N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12131</th>\n",
       "      <td>6666666.139474.787 -&gt; TCA|1|Citrate_synthase_2...</td>\n",
       "      <td>gi|6666666.139474.787|6666666.139474|LFMR01|Ci...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12132</th>\n",
       "      <td>6666666.135434.3838 -&gt; TCA|12|fumarate_hydrata...</td>\n",
       "      <td>gi|6666666.135434.3838|6666666.135434|JYHB01|F...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12133 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       1  \\\n",
       "0      6666666.136315.1646 -> calcium-dependent-antib...   \n",
       "1      6666666.134953.3589 -> polyketide|215|215_1|21...   \n",
       "2      6666666.136574.4344 -> OXALACETATE_AMINOACIDS|...   \n",
       "3      6666666.135439.1862 -> 3PGA_AMINOACIDS|6|Cyste...   \n",
       "4      6666666.136502.4654 -> ALPHAKETOGLUTARATE_AMIN...   \n",
       "...                                                  ...   \n",
       "12128  6666666.102116.464 -> ALPHAKETOGLUTARATE_AMINO...   \n",
       "12129  6666666.94971.5998 -> ALPHAKETOGLUTARATE_AMINO...   \n",
       "12130  6666666.102121.2566 -> ALPHAKETOGLUTARATE_AMIN...   \n",
       "12131  6666666.139474.787 -> TCA|1|Citrate_synthase_2...   \n",
       "12132  6666666.135434.3838 -> TCA|12|fumarate_hydrata...   \n",
       "\n",
       "                                                       2  \n",
       "0      gi|6666666.136315.1646|6666666.136315|AHZY01|A...  \n",
       "1      gi|6666666.134953.3589|6666666.134953|AKCL01|C...  \n",
       "2      gi|6666666.136574.4344|6666666.136574|ATKI01|A...  \n",
       "3      gi|6666666.135439.1862|6666666.135439|JXDG01|C...  \n",
       "4      gi|6666666.136502.4654|6666666.136502|ATLO01|A...  \n",
       "...                                                  ...  \n",
       "12128  gi|6666666.102116.464|6666666.102116|NC_015379...  \n",
       "12129  gi|6666666.94971.5998|6666666.94971|201|Glutam...  \n",
       "12130  gi|6666666.102121.2566|6666666.102121|AJXG01|N...  \n",
       "12131  gi|6666666.139474.787|6666666.139474|LFMR01|Ci...  \n",
       "12132  gi|6666666.135434.3838|6666666.135434|JYHB01|F...  \n",
       "\n",
       "[12133 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pseudo_csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22a82900-13c3-48b6-8c48-2b0fd96bcfd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        gi|6666666.136315.1646|6666666.136315|AHZY01|A...\n",
       "1        gi|6666666.134953.3589|6666666.134953|AKCL01|C...\n",
       "2        gi|6666666.136574.4344|6666666.136574|ATKI01|A...\n",
       "3        gi|6666666.135439.1862|6666666.135439|JXDG01|C...\n",
       "4        gi|6666666.136502.4654|6666666.136502|ATLO01|A...\n",
       "                               ...                        \n",
       "12128    gi|6666666.102116.464|6666666.102116|NC_015379...\n",
       "12129    gi|6666666.94971.5998|6666666.94971|201|Glutam...\n",
       "12130    gi|6666666.102121.2566|6666666.102121|AJXG01|N...\n",
       "12131    gi|6666666.139474.787|6666666.139474|LFMR01|Ci...\n",
       "12132    gi|6666666.135434.3838|6666666.135434|JYHB01|F...\n",
       "Name: 2, Length: 12133, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pseudo_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51244455-dc45-4466-8dbf-61e62f2b6b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_info[1]=pseudo_info[1].replace('6666666.134953','h')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ad5378",
   "metadata": {},
   "source": [
    "<h3>Con las siguientes funciones se depuran los datos y se elabora el data frame de presencia </h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e2856e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_presence(info):\n",
    "    genomes=set()\n",
    "    functions=set()\n",
    "    for name in info:\n",
    "        split=name.split('|')\n",
    "        genomes.add(split[-1])\n",
    "        functions.add(split[4])\n",
    "    presence=pd.DataFrame(index=list(genomes),columns=list(functions))\n",
    "    presence=full_presence(info,presence)\n",
    "    return presence.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49742869",
   "metadata": {},
   "outputs": [],
   "source": [
    "def full_presence(info,presence):\n",
    "    for name in info:\n",
    "        split=name.split('|')\n",
    "        presence.loc[split[-1],split[4]]=1\n",
    "    return presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98496332-0152-41ed-bf57-c918dcb9d81d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_presence=get_presence(pseudo_info)\n",
    "pseudo_presence.sort_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de82cf97-1a22-40d0-9828-2e987ee72066",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b56cf63-13f8-4356-aacf-2a7c3e79f55a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_presence.drop('PseudomonasindicaNBRC103045',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e22f3f7-9309-4bce-bfd4-61f031423f0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "orden=pseudo_presence.sum(axis=1)\n",
    "pseudo_presence['orden']=orden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d3d7094-3980-4344-b90f-50182e56c20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_presence.sort_values(by=['orden'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5e7e880-0b13-4f7b-8c44-9031a2436ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudo_presence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d3bdc4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's assume that \"population\" is a numpy ndarray with your genomes as rows.\n",
    "def calculate_hamming_matrix(population):\n",
    "    # Number of genomes\n",
    "    num_genomes = population.shape[0]\n",
    "    # Create an empty matrix for Hamming distances\n",
    "    hamming_matrix = np.zeros((num_genomes, num_genomes), dtype=int)\n",
    "   # Calculate the Hamming distance between each pair of genomes\n",
    "    for i in range(num_genomes):\n",
    "        for j in range(i+1, num_genomes):  # j=i+1 to avoid calculating the same distance twice\n",
    "            # The Hamming distance is multiplied by the number of genes to convert it into an absolute distance\n",
    "            distance = hamming(population[i], population[j]) * len(population[i])\n",
    "            hamming_matrix[i, j] = distance\n",
    "            hamming_matrix[j, i] = distance  # The matrix is symmetric\n",
    "    \n",
    "    return hamming_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3e242ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_complex(distance_matrix2):\n",
    "    # Create the Rips simplicial complex from the distance matrix\n",
    "    rips_complex = gd.RipsComplex(distance_matrix=distance_matrix2)\n",
    "    # Create the simplex tree from the Rips complex with a maximum dimension of 3\n",
    "    simplex_tree = rips_complex.create_simplex_tree(max_dimension=2)\n",
    "    # Compute the persistence of the simplicial complex\n",
    "    persistence = simplex_tree.persistence()\n",
    "    # Return the persistence diagram or barcode\n",
    "    return persistence, simplex_tree\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19ed433",
   "metadata": {},
   "outputs": [],
   "source": [
    "matriz_pseudo=calculate_hamming_matrix(pseudo_presence.drop(columns=['orden']).values)\n",
    "#matriz_actino=calculate_hamming_matrix(actino_presence.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a15f60cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "matriz_pseudo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06909e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "matriz_pseudo.argmax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee475b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "complex_pseudo=create_complex(matriz_pseudo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c004023",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(matriz_pseudo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd348670",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "gd.plot_persistence_barcode(complex_pseudo[0],legend=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec9a506",
   "metadata": {},
   "outputs": [],
   "source": [
    "gd.plot_persistence_diagram(complex_pseudo[0],legend=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "23f61b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "import gudhi as gd\n",
    "from scipy.spatial.distance import hamming\n",
    "import plotly.graph_objs as go\n",
    "import networkx as nx\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f45bc7ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_simplicial_complex(simplex_tree, filtration_value, vertex_names=None, save_filename=None, plot_size=1, dpi=600, pos=None):\n",
    "    G = nx.Graph()\n",
    "    triangles = []  # List to store triangles (3-nodes simplices)\n",
    "    \n",
    "    for simplex, filt in simplex_tree.get_filtration():\n",
    "        if filt <= filtration_value:\n",
    "            if len(simplex) == 2:\n",
    "                G.add_edge(simplex[0], simplex[1])\n",
    "            elif len(simplex) == 1:\n",
    "                G.add_node(simplex[0])\n",
    "            elif len(simplex) == 3:\n",
    "                triangles.append(simplex)\n",
    "    \n",
    "    # Calculate node positions if not provided\n",
    "    if pos is None:\n",
    "        pos = nx.spring_layout(G)\n",
    "    \n",
    "    # Node trace\n",
    "    x_values, y_values = zip(*[pos[node] for node in G.nodes()])\n",
    "    node_labels = [vertex_names[node] if vertex_names else str(node) for node in G.nodes()]\n",
    "    node_trace = go.Scatter(x=x_values, y=y_values, mode='markers+text', hoverinfo='text', marker=dict(size=14), text=node_labels, textposition='top center', textfont=dict(size=14))\n",
    "    \n",
    "    # Edge traces\n",
    "    edge_traces = []\n",
    "    for edge in G.edges():\n",
    "        x0, y0 = pos[edge[0]]\n",
    "        x1, y1 = pos[edge[1]]\n",
    "        edge_trace = go.Scatter(x=[x0, x1, None], y=[y0, y1, None], mode='lines', line=dict(width=3, color='rgba(0,0,0,0.5)'))\n",
    "        edge_traces.append(edge_trace)\n",
    "    \n",
    "    # Triangle traces\n",
    "    triangle_traces = []\n",
    "    for triangle in triangles:\n",
    "        x0, y0 = pos[triangle[0]]\n",
    "        x1, y1 = pos[triangle[1]]\n",
    "        x2, y2 = pos[triangle[2]]\n",
    "        triangle_trace = go.Scatter(x=[x0, x1, x2, x0, None], y=[y0, y1, y2, y0, None], fill='toself', mode='lines+markers', line=dict(width=2), fillcolor='rgba(255,0,0,0.2)')\n",
    "        triangle_traces.append(triangle_trace)\n",
    "    \n",
    "    # Configure the layout of the plot\n",
    "    layout = go.Layout(showlegend=False, hovermode='closest', xaxis=dict(showgrid=False, zeroline=False, tickfont=dict(size=16, family='Arial, sans-serif')), yaxis=dict(showgrid=False, zeroline=False, tickfont=dict(size=16, family='Arial, sans-serif')))\n",
    "    \n",
    "    fig = go.Figure(data=edge_traces + triangle_traces + [node_trace], layout=layout)\n",
    "    \n",
    "    # Set the figure size\n",
    "    fig.update_layout(width=plot_size * dpi, height=plot_size * dpi)\n",
    "    \n",
    "    # Save the figure if a filename is provided\n",
    "    if save_filename:\n",
    "        pio.write_image(fig, save_filename, width=plot_size * dpi, height=plot_size * dpi, scale=1)\n",
    "    \n",
    "    # Show the figure\n",
    "    #fig.show()\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f2eeaec-7457-44a6-bda7-167a83e2c594",
   "metadata": {},
   "outputs": [],
   "source": [
    "G=visualize_simplicial_complex(complex_pseudo[1],2)\n",
    "nx.write_graphml(G, \"simplex_complex.graphml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993361e7-9c08-486a-b017-76bf7831c1e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from networkx.utils import not_implemented_for, pairwise\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "def minimum_cycle_basis(G, weight=None, total=None):\n",
    "    \"\"\"Returns a minimum weight cycle basis for G\n",
    "\n",
    "    Minimum weight means a cycle basis for which the total weight\n",
    "    (length for unweighted graphs) of all the cycles is minimum.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    G : NetworkX Graph\n",
    "    weight: string\n",
    "        name of the edge attribute to use for edge weights\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    A list of cycle lists.  Each cycle list is a list of nodes\n",
    "    which forms a cycle (loop) in G. Note that the nodes are not\n",
    "    necessarily returned in an order by which they appear in the cycle\n",
    "    \"\"\"\n",
    "    # Get connected components\n",
    "    components = list(nx.connected_components(G))\n",
    "    print(components)\n",
    "    # Use ThreadPoolExecutor to parallelize the computation\n",
    "    with ThreadPoolExecutor() as executor:\n",
    "        # Submit tasks for each connected component\n",
    "        future_to_component = {\n",
    "            executor.submit(_min_cycle_basis, G.subgraph(component), weight, total): component\n",
    "            for component in components\n",
    "        }\n",
    "        \n",
    "        # Collect results as they complete\n",
    "        results = []\n",
    "        for future in as_completed(future_to_component):\n",
    "            component = future_to_component[future]\n",
    "            try:\n",
    "                result = future.result()\n",
    "                results.extend(result)\n",
    "            except Exception as exc:\n",
    "                print(f\"Component {component} generated an exception: {exc}\")\n",
    "\n",
    "    return results\n",
    "\n",
    "def _min_cycle_basis(G, weight,total):\n",
    "    cb = []\n",
    "    cont=0\n",
    "    # We  extract the edges not in a spanning tree. We do not really need a\n",
    "    # *minimum* spanning tree. That is why we call the next function with\n",
    "    # weight=None. Depending on implementation, it may be faster as well\n",
    "    tree_edges = list(nx.minimum_spanning_edges(G, weight=None, data=False))\n",
    "    chords = G.edges - tree_edges - {(v, u) for u, v in tree_edges}\n",
    "\n",
    "    # We maintain a set of vectors orthogonal to sofar found cycles\n",
    "    set_orth = [{edge} for edge in chords]\n",
    "    while set_orth:\n",
    "        if cont==total:\n",
    "            break\n",
    "        base = set_orth.pop()\n",
    "        # kth cycle is \"parallel\" to kth vector in set_orth\n",
    "        cycle_edges = _min_cycle(G, base, weight)\n",
    "        cycle_found=[v for u, v in cycle_edges]\n",
    "\n",
    "        if len(cycle_found)>3:\n",
    "            cb.append(cycle_found)\n",
    "            cont+=1\n",
    "\n",
    "        # now update set_orth so that k+1,k+2... th elements are\n",
    "        # orthogonal to the newly found cycle, as per [p. 336, 1]\n",
    "        set_orth = [\n",
    "            (\n",
    "                {e for e in orth if e not in base if e[::-1] not in base}\n",
    "                | {e for e in base if e not in orth if e[::-1] not in orth}\n",
    "            )\n",
    "            if sum((e in orth or e[::-1] in orth) for e in cycle_edges) % 2\n",
    "            else orth\n",
    "            for orth in set_orth\n",
    "        ]\n",
    "    return cb\n",
    "\n",
    "\n",
    "def _min_cycle(G, orth, weight):\n",
    "    \"\"\"\n",
    "    Computes the minimum weight cycle in G,\n",
    "    orthogonal to the vector orth as per [p. 338, 1]\n",
    "    Use (u, 1) to indicate the lifted copy of u (denoted u' in paper).\n",
    "    \"\"\"\n",
    "    Gi = nx.Graph()\n",
    "\n",
    "    # Add 2 copies of each edge in G to Gi.\n",
    "    # If edge is in orth, add cross edge; otherwise in-plane edge\n",
    "    for u, v, wt in G.edges(data=weight, default=1):\n",
    "        if (u, v) in orth or (v, u) in orth:\n",
    "            Gi.add_edges_from([(u, (v, 1)), ((u, 1), v)], Gi_weight=wt)\n",
    "        else:\n",
    "            Gi.add_edges_from([(u, v), ((u, 1), (v, 1))], Gi_weight=wt)\n",
    "\n",
    "    # find the shortest length in Gi between n and (n, 1) for each n\n",
    "    # Note: Use \"Gi_weight\" for name of weight attribute\n",
    "    spl = nx.shortest_path_length\n",
    "    lift = {n: spl(Gi, source=n, target=(n, 1), weight=\"Gi_weight\") for n in G}\n",
    "\n",
    "    # Now compute that short path in Gi, which translates to a cycle in G\n",
    "    start = min(lift, key=lift.get)\n",
    "    end = (start, 1)\n",
    "    min_path_i = nx.shortest_path(Gi, source=start, target=end, weight=\"Gi_weight\")\n",
    "\n",
    "    # Now we obtain the actual path, re-map nodes in Gi to those in G\n",
    "    min_path = [n if n in G else n[0] for n in min_path_i]\n",
    "\n",
    "    # Now remove the edges that occur two times\n",
    "    # two passes: flag which edges get kept, then build it\n",
    "    edgelist = list(pairwise(min_path))\n",
    "    edgeset = set()\n",
    "    for e in edgelist:\n",
    "        if e in edgeset:\n",
    "            edgeset.remove(e)\n",
    "        elif e[::-1] in edgeset:\n",
    "            edgeset.remove(e[::-1])\n",
    "        else:\n",
    "            edgeset.add(e)\n",
    "\n",
    "    min_edgelist = []\n",
    "    for e in edgelist:\n",
    "        if e in edgeset:\n",
    "            min_edgelist.append(e)\n",
    "            edgeset.remove(e)\n",
    "        elif e[::-1] in edgeset:\n",
    "            min_edgelist.append(e[::-1])\n",
    "            edgeset.remove(e[::-1])\n",
    "\n",
    "    return min_edgelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d137af93",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def find_all_cycles(simplex_tree,persistence):\n",
    "    born_and_number = born_filtraton_value_holes(persistence)\n",
    "    G=nx.Graph()\n",
    "    born=born_and_number.keys()\n",
    "    ciclos_dep=set()\n",
    "    filtration=0\n",
    "    for simplex, filt in simplex_tree.get_filtration():\n",
    "        #if len(ciclos_dep)==num_holes:\n",
    "         #   break\n",
    "        \n",
    "        if filtration!=filt and filtration in born:\n",
    "            number=born_and_number[filtration]\n",
    "            print('se buscan ciclos en el tiempo', filt)\n",
    "            ciclos=minimum_cycle_basis(G,total=number)\n",
    "            for ciclo in ciclos:\n",
    "                if len(ciclo)>3:\n",
    "                    print('Se encontró el ciclo',ciclo,'en el tiempo', filtration)\n",
    "                    ciclos_dep.add(tuple(ciclo))\n",
    "                    #Se llena el hoyo\n",
    "                    for i in ciclo:\n",
    "                        for j in ciclo:\n",
    "                            G.add_edge(i,j)\n",
    "                            \n",
    "            \n",
    "        filtration=filt\n",
    "        \n",
    "        if len(simplex)==2:\n",
    "            G.add_edge(simplex[0], simplex[1])\n",
    "\n",
    "        \n",
    "\n",
    "    return ciclos_dep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ab3799",
   "metadata": {},
   "outputs": [],
   "source": [
    "def born_filtraton_value_holes(persistence):\n",
    "    born=[]\n",
    "    for bar in persistence:\n",
    "        if bar[0]==1:\n",
    "            born.append(bar[1][0])\n",
    "            \n",
    "    born_and_number=set([(x,born.count(x)) for x  in born])\n",
    "\n",
    "    return dict(born_and_number)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dda485c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "ciclos_pseudo=find_all_cycles(complex_pseudo[1],complex_pseudo[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee1f15a7-b857-461e-bf11-f1d80fbceb44",
   "metadata": {},
   "outputs": [],
   "source": [
    "ciclos_pseudo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18bc0dcd-3d16-459f-9de7-a0f7dab664b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from concurrent.futures import ProcessPoolExecutor\n",
    "\n",
    "\n",
    "def find_cycles_at_filtration(G, filtration, born_and_number):\n",
    "    born=born_and_number.keys()\n",
    "    ciclos_dep = set()\n",
    "    if filtration in born:\n",
    "        number=born_and_number[filtration]\n",
    "        print('se buscan',number,'ciclos')\n",
    "        ciclos = minimum_cycle_basis(G,total=20)\n",
    "        print('Se encontraron los ciclos',ciclos)\n",
    "        for ciclo in ciclos:\n",
    "            ciclos_dep.add(tuple(ciclo))\n",
    "                #Se llena el hoyo\n",
    "            for i in ciclo:\n",
    "                for j in ciclo:\n",
    "                    G.add_edge(i, j)\n",
    "    return ciclos_dep\n",
    "\n",
    "def find_all_cycles(simplex_tree, persistence):\n",
    "    born_and_number = born_filtraton_value_holes(persistence)\n",
    "    \n",
    "    G = nx.Graph()\n",
    "    ciclos_dep = set()\n",
    "    filtration = 0\n",
    "    \n",
    "    with ProcessPoolExecutor() as executor:\n",
    "        futures = []\n",
    "        for simplex, filt in simplex_tree.get_filtration():\n",
    "            if filtration != filt:\n",
    "                future = executor.submit(find_cycles_at_filtration, G.copy(), filtration, born_and_number)\n",
    "                futures.append(future)\n",
    "            filtration = filt\n",
    "            if len(simplex) == 2:\n",
    "                G.add_edge(simplex[0], simplex[1])\n",
    "        \n",
    "        for future in futures:\n",
    "            ciclos_dep.update(future.result())\n",
    "\n",
    "    return ciclos_dep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "153b5403-99bd-41f7-9980-683d924348d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "\n",
    "def find_cycle_basis(subgraph):\n",
    "    return list(nx.minimum_cycle_basis(subgraph))\n",
    "\n",
    "def get_subgraphs(graph):\n",
    "    return [graph.subgraph(c) for c in nx.connected_components(graph)]\n",
    "\n",
    "def parallel_minimum_cycle_basis(graph):\n",
    "    subgraphs = get_subgraphs(graph)\n",
    "    with ProcessPoolExecutor() as executor:\n",
    "        results = list(executor.map(find_cycle_basis, subgraphs))\n",
    "    \n",
    "    # Combinar resultados\n",
    "    cycle_basis = [cycle for sublist in results for cycle in sublist]\n",
    "    \n",
    "    return cycle_basis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "131fd7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "[12,12].count(12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af69a131",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing.pool import Pool as ProcessPool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f168436f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ProcessPool(processes=8) as pool:  # context manager providing a `Pool` instance\n",
    "    result = pool.map(find_all_cycles, complex_pseudo[1],complex_pseudo[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7523155",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "contador=0\n",
    "for i in complex_pseudo[0]:\n",
    "    if i[0]==1:\n",
    "        contador+=1\n",
    "contador"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e91f4fbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "G = nx.Graph()\n",
    "nx.add_cycle(G,[195, 212, 167, 187])\n",
    "nx.add_cycle(G,[212, 116, 150, 187])\n",
    "nx.add_cycle(G,[212, 200, 201, 190])\n",
    "\n",
    "#nx.add_cycle(G,[33, 84, 90, 26])\n",
    "#nx.add_cycle(G,[24, 88, 28, 78])\n",
    "#nx.add_cycle(G,[184, 88, 28, 141])\n",
    "#nx.add_cycle(G,[136, 24, 92, 78])\n",
    "\n",
    "#nx.add_cycle(G,[17, 8, 58, 30])\n",
    "#nx.add_cycle(G,[17, 20, 69, 29])\n",
    "#nx.add_cycle(G,[13, 75, 29, 62])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7382ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "nx.draw(G, with_labels = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "320aae59",
   "metadata": {},
   "source": [
    "# complex_pseudo[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2af076b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_cycles(simplex_tree, filtration_value):\n",
    "    G=nx.Graph()\n",
    "    for simplex, filt in simplex_tree.get_filtration():\n",
    "        if filt <= filtration_value:\n",
    "            if len(simplex) == 2:\n",
    "                G.add_edge(simplex[0], simplex[1])\n",
    "    \n",
    "\n",
    "    #ciclos=nx.cycle_basis(G)\n",
    "    ciclos=nx.minimum_cycle_basis(G)\n",
    "    ciclos_dep=set()\n",
    "    for ciclo in ciclos:\n",
    "        if len(ciclo)>3:\n",
    "            print('Se encontró el ciclo',ciclo,'en el tiempo', filtration_value)\n",
    "            ciclos_dep.add(tuple(ciclo))\n",
    "\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64559ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "1+1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80776198",
   "metadata": {},
   "source": [
    "<h2>Generar transferencia horizontal </h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bed0d21e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate(genome, prob, mutated_positions):\n",
    "    \"\"\"Performs mutations on a genome with a certain probability.\"\"\"\n",
    "    mutations = []\n",
    "    for i in range(len(genome)):\n",
    "        if not mutated_positions[i] and np.random.rand() < prob:\n",
    "            genome[i] = 1 if genome[i] == 0 else 0\n",
    "            mutations.append(i)\n",
    "            mutated_positions[i] = True\n",
    "    return genome, mutations\n",
    "\n",
    "def create_generation(population, num_offspring, prob, mutated_positions):\n",
    "    \"\"\"Creates a new generation from the existing population.\"\"\"\n",
    "    new_population = []\n",
    "    new_generation_history = []\n",
    "    for idx, genome in enumerate(population):\n",
    "        for _ in range(num_offspring):\n",
    "            new_genome, mutations = mutate(genome.copy(), prob, mutated_positions)\n",
    "            new_population.append(new_genome)\n",
    "            new_generation_history.append({'parent': idx, 'mutations': mutations})\n",
    "    return new_population, new_generation_history\n",
    "\n",
    "def create_population(num_genes, mutation_probability, num_generations, num_offspring, percentage_ones,semilla):\n",
    "    np.random.seed(semilla)\n",
    "\n",
    "    initial_genome = np.random.choice([0, 1], size=num_genes, p=[1-percentage_ones, percentage_ones])\n",
    "    population = [initial_genome]\n",
    "    genome_history = []\n",
    "    global_mutated_positions = np.zeros(num_genes, dtype=bool)\n",
    "\n",
    "    for generation in range(num_generations):\n",
    "        population, generation_history = create_generation(population, num_offspring, mutation_probability, global_mutated_positions)\n",
    "        genome_history.extend(generation_history)\n",
    "        print(f\"Generation {generation+1}: {len(population)} genomes\")\n",
    "        # Uncomment the next line if you want to see the details of each generation\n",
    "        # print(len(population), generation_history)\n",
    "\n",
    "    return population, genome_history"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb660a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "num_genes = 505\n",
    "mutation_probability = 1 / num_genes\n",
    "num_generations = 7\n",
    "num_offspring = 2\n",
    "percentage_ones = 0.25\n",
    "\n",
    "population, population_genome_history = create_population(num_genes, mutation_probability, num_generations, num_offspring, percentage_ones,42)\n",
    "population = np.array(population)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89a787ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def modify_genomes(population, num_groups, group_size, num_positions_to_change):\n",
    "    np.random.seed(42)  # Puedes usar cualquier número como semilla\n",
    "\n",
    "    # Crear una copia de la población para evitar modificar el original\n",
    "    modified_population = population.copy()\n",
    "\n",
    "    for _ in range(num_groups):\n",
    "        # Seleccionar genomas de manera aleatoria\n",
    "        genome_indices = np.random.choice(population.shape[0], group_size, replace=False)\n",
    "        \n",
    "        # Elegir un punto de inicio al azar y seleccionar posiciones contiguas para cambiar\n",
    "        start_position = np.random.choice(population.shape[1] - num_positions_to_change)\n",
    "        positions_to_change = np.arange(start_position, start_position + num_positions_to_change)\n",
    "        \n",
    "        for index in genome_indices:\n",
    "            # Cambiar los valores en las posiciones seleccionadas a 1\n",
    "            modified_population[index, positions_to_change] = 1\n",
    "    \n",
    "    return modified_population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92862279",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_2=modify_genomes(population, num_groups=2, group_size=3, num_positions_to_change=population.shape[1]//32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "484c695a",
   "metadata": {},
   "outputs": [],
   "source": [
    "hamming_matrix_2 = calculate_hamming_matrix(population_2)\n",
    "#hamming_matrix_2\n",
    "\n",
    "# Print the normalized Hamming distance matrix\n",
    "#print(\"Normalized Hamming Distance Matrix:\")\n",
    "#print(normalized_hamming_matrix_2)\n",
    "persistence2, simplex_tree2 = create_complex(hamming_matrix_2)\n",
    "gd.plot_persistence_barcode(persistence2)\n",
    "gd.plot_persistence_diagram(persistence2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8761fe24",
   "metadata": {},
   "outputs": [],
   "source": [
    "cycles_2=find_all_cycles(simplex_tree2,persistence2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91aa2cb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "contador=0\n",
    "for i in persistence2:\n",
    "    if i[0]==1:\n",
    "        contador+=1\n",
    "contador"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20063d69-9b18-4b2e-b3f1-b6b9ea246503",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73e59b0-9c00-456b-8391-6716a27b44a6",
   "metadata": {},
   "source": [
    "<h1> EJEMPLO</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a781ba3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>UPF0758 protein YeeS (PGF_00064380)</th>\n",
       "      <th>Uncharacterized protein YeeT (PGF_00065498)</th>\n",
       "      <th>YeeU protein (antitoxin to YeeV) (PGF_00067294)</th>\n",
       "      <th>YeeV toxin protein (PGF_00067296)</th>\n",
       "      <th>hypothetical protein (PGF_08225224)</th>\n",
       "      <th>Mobile element protein (PGF_01000105)</th>\n",
       "      <th>Z1226 protein (PGF_00067424)</th>\n",
       "      <th>hypothetical protein (PGF_00111192)</th>\n",
       "      <th>...</th>\n",
       "      <th>hypothetical protein (PGF_04250059)</th>\n",
       "      <th>hypothetical protein (PGF_03025936)</th>\n",
       "      <th>hypothetical protein (PGF_00095896)</th>\n",
       "      <th>hypothetical protein (PGF_06336094)</th>\n",
       "      <th>Conjugal transfer protein traA (PGF_02995735)</th>\n",
       "      <th>hypothetical protein (PGF_05449663)</th>\n",
       "      <th>hypothetical protein (PGF_00284742)</th>\n",
       "      <th>hypothetical protein (PGF_05076506)</th>\n",
       "      <th>hypothetical protein (PGF_03326581)</th>\n",
       "      <th>hypothetical protein (PGF_05267434)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strain</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC00690</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_012952545.1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>193</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC00668</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_012952605.1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>164</td>\n",
       "      <td>49</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC00609</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_012952615.1</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>189</td>\n",
       "      <td>66</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC00701</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_012952535.1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>181</td>\n",
       "      <td>58</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC00678</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>GCF_012952555.1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>142</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00215</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952415.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>221</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00221</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952335.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00218</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952365.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>251</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00213</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952385.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>219</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00187</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952405.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>109</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00149</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>GCF_012952515.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLP00177</th>\n",
       "      <td>Klebsiella quasipneumoniae</td>\n",
       "      <td>GCF_012952465.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 13167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Taxonomy           RefSeq  \\\n",
       "Strain                                                  \n",
       "EC00690             Escherichia coli  GCF_012952545.1   \n",
       "EC00668             Escherichia coli  GCF_012952605.1   \n",
       "EC00609             Escherichia coli  GCF_012952615.1   \n",
       "EC00701             Escherichia coli  GCF_012952535.1   \n",
       "EC00678             Escherichia coli  GCF_012952555.1   \n",
       "KLP00215       Klebsiella pneumoniae  GCF_012952415.1   \n",
       "KLP00221       Klebsiella pneumoniae  GCF_012952335.1   \n",
       "KLP00218       Klebsiella pneumoniae  GCF_012952365.1   \n",
       "KLP00213       Klebsiella pneumoniae  GCF_012952385.1   \n",
       "KLP00187       Klebsiella pneumoniae  GCF_012952405.1   \n",
       "KLP00149       Klebsiella pneumoniae  GCF_012952515.1   \n",
       "KLP00177  Klebsiella quasipneumoniae  GCF_012952465.1   \n",
       "\n",
       "          UPF0758 protein YeeS (PGF_00064380)  \\\n",
       "Strain                                          \n",
       "EC00690                                     4   \n",
       "EC00668                                     3   \n",
       "EC00609                                     7   \n",
       "EC00701                                     3   \n",
       "EC00678                                     6   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          Uncharacterized protein YeeT (PGF_00065498)  \\\n",
       "Strain                                                  \n",
       "EC00690                                             4   \n",
       "EC00668                                             3   \n",
       "EC00609                                             6   \n",
       "EC00701                                             3   \n",
       "EC00678                                             6   \n",
       "KLP00215                                            0   \n",
       "KLP00221                                            0   \n",
       "KLP00218                                            0   \n",
       "KLP00213                                            0   \n",
       "KLP00187                                            0   \n",
       "KLP00149                                            0   \n",
       "KLP00177                                            0   \n",
       "\n",
       "          YeeU protein (antitoxin to YeeV) (PGF_00067294)  \\\n",
       "Strain                                                      \n",
       "EC00690                                                 4   \n",
       "EC00668                                                 3   \n",
       "EC00609                                                 5   \n",
       "EC00701                                                 4   \n",
       "EC00678                                                 6   \n",
       "KLP00215                                                0   \n",
       "KLP00221                                                0   \n",
       "KLP00218                                                0   \n",
       "KLP00213                                                0   \n",
       "KLP00187                                                0   \n",
       "KLP00149                                                0   \n",
       "KLP00177                                                0   \n",
       "\n",
       "          YeeV toxin protein (PGF_00067296)  \\\n",
       "Strain                                        \n",
       "EC00690                                   4   \n",
       "EC00668                                   2   \n",
       "EC00609                                   5   \n",
       "EC00701                                   3   \n",
       "EC00678                                   5   \n",
       "KLP00215                                  0   \n",
       "KLP00221                                  0   \n",
       "KLP00218                                  0   \n",
       "KLP00213                                  0   \n",
       "KLP00187                                  0   \n",
       "KLP00149                                  0   \n",
       "KLP00177                                  0   \n",
       "\n",
       "          hypothetical protein (PGF_08225224)  \\\n",
       "Strain                                          \n",
       "EC00690                                   193   \n",
       "EC00668                                   164   \n",
       "EC00609                                   189   \n",
       "EC00701                                   181   \n",
       "EC00678                                   142   \n",
       "KLP00215                                  221   \n",
       "KLP00221                                   16   \n",
       "KLP00218                                  251   \n",
       "KLP00213                                  219   \n",
       "KLP00187                                  109   \n",
       "KLP00149                                  175   \n",
       "KLP00177                                  105   \n",
       "\n",
       "          Mobile element protein (PGF_01000105)  Z1226 protein (PGF_00067424)  \\\n",
       "Strain                                                                          \n",
       "EC00690                                      63                             3   \n",
       "EC00668                                      49                             1   \n",
       "EC00609                                      66                             4   \n",
       "EC00701                                      58                             2   \n",
       "EC00678                                      20                             4   \n",
       "KLP00215                                     21                             0   \n",
       "KLP00221                                     30                             0   \n",
       "KLP00218                                     11                             0   \n",
       "KLP00213                                     27                             0   \n",
       "KLP00187                                     27                             0   \n",
       "KLP00149                                     20                             0   \n",
       "KLP00177                                     17                             0   \n",
       "\n",
       "          hypothetical protein (PGF_00111192)  ...  \\\n",
       "Strain                                         ...   \n",
       "EC00690                                     0  ...   \n",
       "EC00668                                     0  ...   \n",
       "EC00609                                     0  ...   \n",
       "EC00701                                     0  ...   \n",
       "EC00678                                     2  ...   \n",
       "KLP00215                                    0  ...   \n",
       "KLP00221                                    0  ...   \n",
       "KLP00218                                    0  ...   \n",
       "KLP00213                                    0  ...   \n",
       "KLP00187                                    0  ...   \n",
       "KLP00149                                    0  ...   \n",
       "KLP00177                                    0  ...   \n",
       "\n",
       "          hypothetical protein (PGF_04250059)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_03025936)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_00095896)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_06336094)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          Conjugal transfer protein traA (PGF_02995735)  \\\n",
       "Strain                                                    \n",
       "EC00690                                               0   \n",
       "EC00668                                               0   \n",
       "EC00609                                               0   \n",
       "EC00701                                               0   \n",
       "EC00678                                               0   \n",
       "KLP00215                                              0   \n",
       "KLP00221                                              0   \n",
       "KLP00218                                              0   \n",
       "KLP00213                                              0   \n",
       "KLP00187                                              0   \n",
       "KLP00149                                              0   \n",
       "KLP00177                                              0   \n",
       "\n",
       "          hypothetical protein (PGF_05449663)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_00284742)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_05076506)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_03326581)  \\\n",
       "Strain                                          \n",
       "EC00690                                     0   \n",
       "EC00668                                     0   \n",
       "EC00609                                     0   \n",
       "EC00701                                     0   \n",
       "EC00678                                     0   \n",
       "KLP00215                                    0   \n",
       "KLP00221                                    0   \n",
       "KLP00218                                    0   \n",
       "KLP00213                                    0   \n",
       "KLP00187                                    0   \n",
       "KLP00149                                    0   \n",
       "KLP00177                                    0   \n",
       "\n",
       "          hypothetical protein (PGF_05267434)  \n",
       "Strain                                         \n",
       "EC00690                                     0  \n",
       "EC00668                                     0  \n",
       "EC00609                                     0  \n",
       "EC00701                                     0  \n",
       "EC00678                                     0  \n",
       "KLP00215                                    0  \n",
       "KLP00221                                    0  \n",
       "KLP00218                                    0  \n",
       "KLP00213                                    0  \n",
       "KLP00187                                    0  \n",
       "KLP00149                                    0  \n",
       "KLP00177                                    0  \n",
       "\n",
       "[12 rows x 13167 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "link=\"https://raw.githubusercontent.com/shadayguerrero/TDA_Horizontal_Genes_Transfer/main/data/BVBRC_protein_families_all_final.csv\"\n",
    "short=pd.read_csv(link,index_col=0)\n",
    "short\n",
    "# Re-index for the Strain\n",
    "df_reindexed = short.set_index('Strain', inplace=False)\n",
    "df_reindexed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff38a193-b33b-45e4-a810-6d2b3ec4e3e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "123"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names_to_search = [\n",
    "    \"PGF_04886779\", \"PGF_04370656\", \"PGF_07342234\", \"PGF_03299821\", \"PGF_05419346\",\n",
    "    \"PGF_03084064\", \"PGF_04563108\", \"PGF_00414310\", \"PGF_00022105\", \"PGF_05419346\",\n",
    "    \"PGF_06607596\", \"PGF_00047661\", \"PGF_00216543\", \"PGF_08225224\", \"PGF_08225224\",\n",
    "    \"PGF_00110526\", \"PGF_00355535\", \"PGF_01650562\", \"PGF_04243824\", \"PGF_02333019\",\n",
    "    \"PGF_08225224\", \"PGF_03145724\", \"PGF_06495825\", \"PGF_00452501\", \"PGF_05493974\",\n",
    "    \"PGF_00020282\", \"PGF_06958254\", \"PGF_09224510\", \"PGF_07563198\", \"PGF_01000105\",\n",
    "    \"PGF_01031760\", \"PGF_01569843\", \"PGF_01077017\", \"PGF_01000105\", \"PGF_02752571\",\n",
    "    \"PGF_00410347\", \"PGF_01958521\", \"PGF_08225224\", \"PGF_07207755\", \"PGF_00010376\",\n",
    "    \"PGF_03543440\", \"PGF_05156704\", \"PGF_01000105\", \"PGF_01000105\", \"PGF_02932429\",\n",
    "    \"PGF_06674747\", \"PGF_07498135\", \"PGF_10469466\", \"PGF_08225224\", \"PGF_01958554\",\n",
    "    \"PGF_05971953\", \"PGF_00904239\", \"PGF_00424315\", \"PGF_08238808\", \"PGF_08225224\",\n",
    "    \"PGF_08238808\", \"PGF_08225224\", \"PGF_00013862\", \"PGF_00013864\", \"PGF_10387734\",\n",
    "    \"PGF_09290469\", \"PGF_08225224\", \"PGF_00013946\", \"PGF_01949998\", \"PGF_00013916\",\n",
    "    \"PGF_08972330\", \"PGF_08225224\", \"PGF_00013925\", \"PGF_00013923\", \"PGF_03830470\",\n",
    "    \"PGF_00326234\", \"PGF_00163448\", \"PGF_00013943\", \"PGF_02913871\", \"PGF_00013942\",\n",
    "    \"PGF_00615364\", \"PGF_00335754\", \"PGF_07717077\", \"PGF_00013945\", \"PGF_00013871\",\n",
    "    \"PGF_03332290\", \"PGF_08225224\", \"PGF_06611914\", \"PGF_00089171\", \"PGF_06560088\",\n",
    "    \"PGF_00106691\", \"PGF_04419840\", \"PGF_00467825\", \"PGF_00013893\", \"PGF_00013875\",\n",
    "    \"PGF_01246464\", \"PGF_00013866\", \"PGF_08225224\", \"PGF_00859415\", \"PGF_00013863\",\n",
    "    \"PGF_00411196\", \"PGF_00326542\", \"PGF_00064344\", \"PGF_07396822\", \"PGF_07978541\",\n",
    "    \"PGF_01595863\", \"PGF_06275535\", \"PGF_06275535\", \"PGF_01950491\", \"PGF_01958592\",\n",
    "    \"PGF_01009331\", \"PGF_00038929\", \"PGF_00038932\", \"PGF_05682743\", \"PGF_00064362\",\n",
    "    \"PGF_02142873\", \"PGF_04180579\", \"PGF_00072490\", \"PGF_07517035\", \"PGF_00324174\",\n",
    "    \"PGF_00106708\", \"PGF_10494469\", \"PGF_08970263\", \"PGF_08225224\", \"PGF_00060224\", \n",
    "    \"PGF_00415171\", \"PGF_07518314\", \"PGF_00424726\"]\n",
    "selected_columns = [col for col in df_reindexed.columns if any(word in col for word in names_to_search)]\n",
    "column_index = [df_reindexed.columns.get_loc(col) for col in selected_columns]\n",
    "\n",
    "# Create a new DataFrame with only the selected columns\n",
    "df_filtered = df_reindexed[selected_columns]\n",
    "\n",
    "len(names_to_search)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1ee77bdc-4fc2-4187-9d10-b3909e3f8327",
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_names=list(df_reindexed.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "806d80f1-cba1-4dc5-871e-15907ab38213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 1, 1, ..., 0, 2, 0],\n",
       "       [1, 1, 1, ..., 0, 3, 0],\n",
       "       [0, 1, 1, ..., 0, 1, 0],\n",
       "       ...,\n",
       "       [0, 1, 1, ..., 0, 0, 0],\n",
       "       [0, 1, 1, ..., 0, 2, 2],\n",
       "       [0, 1, 1, ..., 0, 0, 0]])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#PatientA_B=all.iloc[:,4438:4520].values\n",
    "PatientA_B=df_reindexed.iloc[:,4020:4520].values\n",
    "PatientA_B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8654dd6-300f-4dc7-8f2c-0705bcd2e0eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1, (118.0, 145.0)),\n",
       " (1, (88.0, 112.0)),\n",
       " (0, (0.0, inf)),\n",
       " (0, (0.0, 108.0)),\n",
       " (0, (0.0, 85.0)),\n",
       " (0, (0.0, 72.0)),\n",
       " (0, (0.0, 61.0)),\n",
       " (0, (0.0, 53.0)),\n",
       " (0, (0.0, 43.0)),\n",
       " (0, (0.0, 31.0)),\n",
       " (0, (0.0, 30.0)),\n",
       " (0, (0.0, 18.0)),\n",
       " (0, (0.0, 18.0)),\n",
       " (0, (0.0, 18.0))]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hamming_matrix_3 = calculate_hamming_matrix(PatientA_B)\n",
    "persistence3, simplex_tree3 = create_complex(hamming_matrix_3)\n",
    "persistence3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "39a80661-dbff-44de-a394-e5cd7174c7f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usetex mode requires TeX.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:title={'center':'Persistence barcode'}>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 100x100 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAG3CAYAAAAU+jfPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAyKklEQVR4nO3de3RU1f3//9fkNkkgFxIkYSBcRUFQUAIoKASq2Aio9VMrXgC1ro8oCogfIZRWAT8SQMUICH5gWbQiSrWAaLUalWutCgmg4gVoERDBKJck3JKQ7N8f/nK+DJkEEmb2MOH5WGtW15yzz9nvPUMzL/ecs8dljDECAACwJCzYBQAAgHML4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDCIJWrVrJ5XLpu+++C3Yp9U5GRoZcLpdWrlwZ7FLOKitXrpTL5VJGRkawSwEIHwhdlR/gJz5iYmLUtm1b3X333dq8eXOwSwy4iRMnauLEicEuAwBqJSLYBQBnql27dmrSpIkk6eDBg9q6dasWLFigRYsW6fXXX9egQYOCXGFVbdu2VXR0tCIjI8/oPJMmTZIkAgiAkEL4QMj7wx/+oDvvvNN5/uOPP+qOO+7QBx98oLvuukvfffedGjZsGLwCffjwww+DXQIABA1fu6DeSUlJ0csvvyy32619+/YpNzc32CUBAE5A+EC9lJqaqnbt2kmStm7d6rXvm2++0d13361WrVrJ7XYrOTlZAwYM0EcffeTzXCdeHLpixQplZmaqcePGXhc1Hj9+XM8++6y6d++uuLg4ud1ueTwe9ezZU4899pgOHjxY7TlPdPjwYU2ePFmXXHKJGjRooOjoaKWlpSkjI0NTp05VWVmZpF++ZnG5XM5xJ1/7cvJ5v//+e40cOVIXXHCBYmJilJiYqL59++qNN97wOeYTL9r85ptvdPPNN6tx48aKiYlR165d9de//rWml1+5ubm66aab5PF4nNeib9++eu6551RSUlKlfW3fk9P12WefacCAAUpKSlKDBg3Us2dPLVu2zGfbgwcP6oUXXtANN9yg888/XzExMUpISFCPHj00c+ZMHT9+3Odxla+5JP3tb39T7969lZiYWOV92L9/vx577DFdeumlio+PV8OGDdWhQwcNHz5cGzZsqHLeffv2aezYsbrwwgsVExOjRo0aKSMjQ6+88oqMMdWOeenSperZs6caNGig5ORkDRw4UOvXrz/la7V//35NmDBBnTp1UoMGDRQXF6fLL79c8+fPV0VFxSmPB2rFACGqZcuWRpJZsGCBz/0dO3Y0ksy0adOcbYsXLzZRUVFGkomLizNdunQxqampRpJxuVxm5syZ1fYzZcoUExYWZho1amS6detmmjdvblasWGGMMea//uu/jCQjybRt29Z069bNpKWlmfDwcCPJbNiwwec5t2/f7mwrKyszl19+uZFkwsLCzIUXXmjS09ONx+MxYWFhRpI5cOCAMcaYF154wfTq1cvps1evXl6PPXv2OOdduXKlSUhIMJJMTEyMufjii01aWppz7MMPP1xlzH369DGSzFNPPWUaNmxo4uLiTNeuXc15553nHPfyyy/7fN1HjBjhtElOTjbp6emmZcuWzhhOHHNd35OaVNY+efJkExUVZRo2bGjS09NN06ZNnbqefvrpKse9/PLLRpKJiooyLVu2NN26dTNt2rRx6h4wYIApLy+vclzlOadOnWokmZSUFNOtWzdz3nnnOWPduHGj8Xg8znt70UUXmS5dupj4+HgjyQwbNszrnFu3bnXeo6ioKHPZZZeZNm3aOH0NHTrUVFRUVKll2rRpTpumTZuarl27moYNGxq3220ef/xxI8n06dOnynFffvmladasmdPfRRddZNq2bWtcLpeRZH7729/67A+oK8IHQlZN4WPPnj3G7XYbSeZvf/ubMcaYTZs2GbfbbaKjo828efO8PkiWL19u4uPjTXh4uNm4caPPfsLDw82kSZNMWVmZMcaYiooKc+zYMbN+/XojyaSlpZmvvvrK69jCwkIzf/58s3PnTp/nPPGD+I033jCSTOfOnc2uXbu82hcUFJicnBxz+PBhr+2VHzTV2b17t0lKSjIul8tMmTLFHDt2zNn3z3/+0/nAeeutt7yOq/wAj4yMNA888IA5evSoM+Zx48YZScbj8Zjjx497HZeTk2MkmdjYWPPyyy97vcb79u0zTz/9tCkoKHC21fU9qUll7REREWbw4MHm0KFDTu0zZ8509p18zk2bNpm3337b6zUyxph///vfpnfv3kaSefHFF6v0V/keREVFmXnz5jkf0mVlZaasrMwUFhaaFi1aGEnm17/+dZX3dvXq1WbhwoXO84qKCpOenu4Ehb179zr73n33XdOgQQMjycyZM8frPPn5+SY8PNy4XC4ze/Zsp47i4mJzyy23mMjISJ/h49ChQ6Zt27ZGkhk5cqQpLCx09m3evNkJ8bNnz67xdQdqg/CBkFVd+Pjxxx/N1VdfbSSZRo0amaKiImOMMTfddJORZJ599lmf55s1a5aRZO6++26f/QwaNMjnca+++qqRZB566KFa135i+MjOzq6xPl9OFT7GjBlTY21vvfWWkWT69evntb3yA7xz585V/mu/tLTUmZnIz893th85csQkJycbSeYvf/nLadVf1/ekJpW1N2nSxAlNvvocOnToaZ9z27ZtRpK55pprquyrfA8efPBBn8dOnz7dSDIdOnSoEmx8yc3NNZKM2+32msE6+XwtW7b0mo244447jCRz8803Vznm6NGjpkmTJj7DR2Ug+81vfuOznk2bNhmXy2XatGlzytqB00X4QMiq/ABv166d83VDx44dnSn8yMhI88YbbxhjjCkpKTHR0dEmPDzc67/sTrRjxw7naxNf/bz++us+j1u7dq2RZC655BKzb9++WtV+YvhYuHChkWT69+9fZYajOqcKH61atTKSzObNm33uLysrM1FRUSY6OtqZ0THm/32AVxcKrr32WiPJeX2N+X8fmh6Px+fXEyc7k/ekJpW1jx8/3uf+Dz74wPl65GTHjh0zr7zyirnnnntM//79zZVXXun826oMNCerfA/WrVvns7/KY59//vnTqv+RRx4xkswdd9zhc39RUZEzq/f111872yvDxUcffeTzuKysLJ/hIyMjw0gyf//736utqXXr1kZSlVkboK641RYhb+vWrc5FpVFRUUpNTVXv3r318MMPq0uXLpKkLVu26NixY4qKitJ1113n8zzm/7+Ib/fu3T73d+jQwef2K664Qj169NCnn36qtLQ0XXPNNerdu7f69Omjyy67zOvC0JrceOONatWqld5//315PB79+te/1lVXXaWMjAx17NjxtM5xokOHDjkXPP73f/93jW2PHTumffv2KSUlxWt727ZtfbavXFfl0KFDzravv/5aktS9e3eFhZ36WnZ/vCc1qe79qtz+448/qqioSPHx8ZKknTt3qn///vr222+rPef+/ftr3V/l63L55ZefVt1btmyRJF100UU+98fFxSktLU3btm3Tli1b1L59ex08eFAFBQU11lHd9i+++EKS9Oijj2rKlCk+2/z888+SfnkfmjdvflrjAGpC+EDIW7Bggdc6H74UFhZKkkpLS/XPf/6zxrbHjh3zub1BgwY+t4eFhendd9/VpEmTtHDhQr355pt68803JUktW7bUxIkTT1lf5fnXrFmjRx99VG+88YYWL16sxYsXS/rlg2jatGkaOHDgKc9TqXLMkk45Zkk6evSoz5p8qQwXleFAkoqKiiRJiYmJtarvTN6TmlQGpJq2FxcXO+Hjzjvv1LfffqsePXpo0qRJ6tKli5KSkhQZGanjx487/1ud6l6r2r4ulYGuuvqlX24n37Ztm4qLi72OkaTzzjuv2mN8qXwf8vLyTlmbr38jQF1wqy3OCZWLjDVr1kzml68ba3zUVqNGjZSTk6OffvpJGzZs0LPPPqu+fftqx44duuuuu6q9pfVkzZs315///Gft379fn3zyiaZOnar09HR99dVXuvHGG/Xpp5/WeszSLx/wpxpzq1atajtsL3FxcZJU5bbiU9UXqPfkp59+OuX2ypp/+OEHrVixQrGxsXrnnXd07bXXKiUlxVmBdteuXbXu/+Q+avu6VM5k+PLjjz96nfvE97q6cVd3vspjt27desr3gN+Fgb8QPnBOaNeunSIjI7Vnz54ap87PlMvlUpcuXTRy5Eh99NFHysrKkiTNnz+/VueJiIhQjx49NG7cOK1bt06DBw9WeXm5/vznP5/2ORISEuTxeCTJyu/cVH41tG7dutNaFyLQ70nl1x3VbU9JSXFmPXbs2CFJat++vZKSkqocs2nTpjrXUfm6fPLJJ6fV/oILLpAkffXVVz73FxcXO2Gosm1iYqIzU/LNN9/4PK6616Py650vv/zytOoD/IHwgXNCbGysrr32WlVUVGjmzJnW+q38nv+HH34IyHliYmIkVT8dftNNN0mScnJyzqj/09GrVy81btxYu3fv1quvvnrK9oF+T1544QWfC5rNmTNHktS/f39nW+XrWFBQ4HOWZfr06XWu48Ybb5QkzZo1S6Wlpadsf+2110qSXn/9de3du7fK/v/7v/9TSUmJWrZsqQsvvNDZfs0110iSnn/++SrHlJSUVBtcK/+NzJw5s04zTECdBPBiViCgTrXI2Mk2bNhg3G63CQ8PN9nZ2ebIkSNe+3/44QeTk5Nj5s6d67OfkxfHqrRw4UIzefLkKvt//vln069fP5+3dfo654wZM8wzzzzjta6DMb/c8dGpUycjyTz66KNe+yrXYHj33Xd91rZr1y6TlJTk3G5buUhZpX379pkXXnjBPP74417bK+8YqVxE7WTDhg3z+do/++yzRpJp0KCBWbRokdetoPv37zczZszwWuejru9JTU5c5+P222/3WufjueeeMy6Xy4SHh3st/FZaWmoaNWpkJJnHH3/cqfvo0aNm5MiRJjo6uto7i6rbXqmoqMh5vwcMGGC+//57r/1r1qypss5Ht27djCSTkZFhfvzxR2ffe++9Zxo2bGgkVXlN1q9fb8LCwozL5TJz5851xnDo0CFz2223VbvOR3FxsbOA2a233mp++OGHKvsXL15cq1vJgVMhfCBk1TZ8GGPMkiVLTGxsrJFkoqOjTZcuXUz37t29VvwcN26cz36qCx/PPPOMc2yzZs1Mt27dTKdOnZxbfps1a2Z27NhxynOOGjXKOU+rVq1M9+7dTfv27Z1VUjt16mQOHjzodZ7Jkyc7C6Bdeumlpk+fPqZPnz5e60OsXbvWNG7c2Ln9+OKLLzY9evQwbdq0cVawvOWWW7zOW9fwUVFRYe677z5nHI0bNzbdunUzrVq1csZx8utYl/ekJievcBoXF+esFFt5vunTp1c5bvbs2c7+1NRUk56ebuLj443L5TLz58+vc/gw5pcVTivXRgkLCzMdO3Y0Xbp0cVae9bXCafPmzZ31Pi677DJz/vnnO30NGTLE54qjU6ZMcdp4PB6Tnp5u4uLiTrnC6ddff+3cThsWFmY6dOhgevToYS644ALnfevRo0fNLzxQC4QPhKy6hA9jjPnuu+/MqFGjTPv27U1MTIxp2LChufDCC81vfvMb89JLL1WZHThV+Ni5c6eZNm2aueaaa0yLFi1MdHS0SU5ONpdddpn53//93yrnq+6cX3/9tZk4caLp3bu3adasmYmKijIpKSnm8ssvN7NmzaoyK2DML//F/thjj5kLL7zQWfvBV60FBQVmwoQJpnPnzqZhw4YmJibGnH/++SYzM9PMmTOnymxLXcNHpb///e9m4MCB5rzzzjNRUVGmWbNmpl+/fmbOnDmmpKSkSvvavic1ObH2Tz/91GRmZprExEQTExNjLr/8crNkyZJqj124cKHp0qWLiYqKMomJiaZfv37OrNKZhA9jfpkJGz9+vOnYsaOJjY01cXFxpkOHDua+++7zuYLrTz/9ZP7nf/7HtGvXzrjdbhMfH2969+5tXn755RqXOn/jjTdMjx49TExMjGnUqJG57rrrzLp168yKFSuqDR/G/DJDM3XqVNOjRw8THx9v3G63adWqlenXr5956qmnqv33D9SFyxi+5AMAAPZwwSkAALCK8AEAAKwifAAAAKtqHT5Wr16tQYMGyePxyOVyadmyZdW2vffee+VyuaysMQAAAEJDrcPH4cOH1blzZ82ePbvGdsuWLdOnn37qrLAIAAAg1eGH5TIzM5WZmVljm927d+uBBx7Qe++9pwEDBtS5OAAAUP/4/VdtKyoqNGTIED3yyCOn9TPgJSUlXksgV1RUaP/+/UpOTj7tnyIHAADBZYxRcXGxPB6P88vX1fF7+Jg2bZoiIiI0cuTI02qfnZ2tSZMm+bsMAAAQBLt27VLz5s1rbOPX8JGXl6dnn31W+fn5pz1rMX78eI0ZM8Z5XlhYqBYtWmjXrl3OL04CAICzW1FRkdLS0hQXF3fKtn4NH2vWrFFBQYFatGjhbCsvL9fDDz+snJwcfffdd1WOcbvdcrvdVbbHx8cTPgAACDGnM/ng1/AxZMgQXX311V7brr32Wg0ZMkR33XWXP7sCAAAhqtbh49ChQ9q2bZvzfPv27dq4caOSkpLUokULJScne7WPjIxUamqqLrzwwjOvFgAAhLxah4/169erb9++zvPK6zWGDRumF1980W+FAQCA+qnW4SMjI0O1+SFcX9d5AACAcxe/7QIAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACs8usPy51rDozNCnYJAACcFYpKSk67LTMfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKtcxhgT7CJOVFRUpISEBBUWFio+Pj7Y5QAAgNNQm89vZj4AAIBVhA8AAGAV4QMAAFhF+AAAAFZFBLuAUHZgbFawSwAAv2g0fWqwS8A5hJkPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhV6/CxevVqDRo0SB6PRy6XS8uWLXP2lZWVady4cbr44ovVoEEDeTweDR06VD/88IM/awYAACGs1uHj8OHD6ty5s2bPnl1l35EjR5Sfn68//elPys/P15IlS7RlyxZdf/31fikWAACEvlovMpaZmanMzEyf+xISEpSbm+u1bdasWerevbt27typFi1aVDmmpKREJSUlzvOioqLalgQAAEJIwK/5KCwslMvlUmJios/92dnZSkhIcB5paWmBLgkAAARRQMPHsWPHlJWVpdtuu03x8fE+24wfP16FhYXOY9euXYEsCQAABFnAftulrKxMgwcPVkVFhebMmVNtO7fbLbfbHagyAADAWSYg4aOsrEy/+93vtH37dn300UfVznoAAIBzj9/DR2Xw2Lp1q1asWKHk5GR/dwEAAEJYrcPHoUOHtG3bNuf59u3btXHjRiUlJcnj8ei3v/2t8vPz9fbbb6u8vFx79+6VJCUlJSkqKsp/lQMAgJBU6/Cxfv169e3b13k+ZswYSdKwYcM0ceJELV++XJLUpUsXr+NWrFihjIyMulcKAADqhVqHj4yMDBljqt1f0z4AAAB+2wUAAFhF+AAAAFYRPgAAgFUBW2TsXNBo+tRglwAAQMhh5gMAAFhF+AAAAFYRPgAAgFWEDwAAYBUXnJ6BA2Ozgl0CgDPEheOAfcx8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKyKCHYBoazR9KnBLgEAgJDDzAcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIq7Xc7AgbFZwS4BCCncIQZAYuYDAABYRvgAAABWET4AAIBVhA8AAGAV4QMAAFhV6/CxevVqDRo0SB6PRy6XS8uWLfPab4zRxIkT5fF4FBMTo4yMDG3evNlf9QIAgBBX6/Bx+PBhde7cWbNnz/a5f/r06ZoxY4Zmz56tdevWKTU1Vddcc42Ki4vPuFgAABD6ar3OR2ZmpjIzM33uM8YoJydHEyZM0E033SRJeumll5SSkqJFixbp3nvvPbNqAQBAyPPrNR/bt2/X3r171b9/f2eb2+1Wnz599PHHH/s8pqSkREVFRV4PAABQf/k1fOzdu1eSlJKS4rU9JSXF2Xey7OxsJSQkOI+0tDR/lgQAAM4yAbnbxeVyeT03xlTZVmn8+PEqLCx0Hrt27QpESQAA4Czh1992SU1NlfTLDEjTpk2d7QUFBVVmQyq53W653W5/lgEAAM5ifp35aN26tVJTU5Wbm+tsKy0t1apVq9SzZ09/dgUAAEJUrWc+Dh06pG3btjnPt2/fro0bNyopKUktWrTQ6NGjNWXKFLVr107t2rXTlClTFBsbq9tuu82vhQMAgNBU6/Cxfv169e3b13k+ZswYSdKwYcP04osvauzYsTp69Kjuv/9+HThwQD169ND777+vuLg4/1UNAABClssYY4JdxImKioqUkJCgwsJCxcfHB7ucGh0YmxXsEoCQ0mj61GCXACBAavP5zW+7AAAAqwgfAADAKsIHAACwivABAACs8usiY+caLp4DAKD2mPkAAABWET4AAIBVhA8AAGAV4QMAAFjFBadngBVOEYq4UBpAsDHzAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALAqItgFhLJG06cGuwQAAEIOMx8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrudjkDB8ZmBbsE1FPcSQWgPmPmAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABY5ffwcfz4cf3xj39U69atFRMTozZt2mjy5MmqqKjwd1cAACAE+f1W22nTpun555/XSy+9pI4dO2r9+vW66667lJCQoFGjRvm7OwAAEGL8Hj7+9a9/6YYbbtCAAQMkSa1atdKrr76q9evX+7srAAAQgvz+tcuVV16pDz/8UFu2bJEkbdq0SWvXrtV1113ns31JSYmKioq8HgAAoP7y+8zHuHHjVFhYqPbt2ys8PFzl5eV64okndOutt/psn52drUmTJvm7DAAAcJby+8zH4sWLtXDhQi1atEj5+fl66aWX9NRTT+mll17y2X78+PEqLCx0Hrt27fJ3SQAA4Czi95mPRx55RFlZWRo8eLAk6eKLL9aOHTuUnZ2tYcOGVWnvdrvldrv9XQYAADhL+X3m48iRIwoL8z5teHg4t9oCAABJAZj5GDRokJ544gm1aNFCHTt21IYNGzRjxgzdfffd/u4KAACEIL+Hj1mzZulPf/qT7r//fhUUFMjj8ejee+/Vo48+6u+uAABACPJ7+IiLi1NOTo5ycnL8fWoAAFAP8NsuAADAKsIHAACwivABAACsInwAAACr/H7B6bmk0fSpwS4BAICQw8wHAACwivABAACsInwAAACrCB8AAMAqLjg9AwfGZgW7BFjAhcUA4F/MfAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsInwAAACrCB8AAMAqwgcAALCK8AEAAKwifAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKsIHAACwivABAACsigh2AaGs0fSpwS4BAICQw8wHAACwivABAACsInwAAACrCB8AAMAqLjg9AwfGZgW7hHqNC3oBoH5i5gMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVgUkfOzevVt33HGHkpOTFRsbqy5duigvLy8QXQEAgBDj93U+Dhw4oF69eqlv375699131aRJE/373/9WYmKiv7sCAAAhyO/hY9q0aUpLS9OCBQucba1atfJ3NwAAIET5/WuX5cuXKz09XTfffLOaNGmiSy+9VPPnz6+2fUlJiYqKirweAACg/vJ7+PjPf/6juXPnql27dnrvvfc0fPhwjRw5Un/5y198ts/OzlZCQoLzSEtL83dJAADgLOIyxhh/njAqKkrp6en6+OOPnW0jR47UunXr9K9//atK+5KSEpWUlDjPi4qKlJaWpsLCQsXHx/uzNL/jt10Ci992AYDQUVRUpISEhNP6/Pb7zEfTpk110UUXeW3r0KGDdu7c6bO92+1WfHy81wMAANRffg8fvXr10rfffuu1bcuWLWrZsqW/uwIAACHI7+HjoYce0ieffKIpU6Zo27ZtWrRokebNm6cRI0b4uysAABCC/B4+unXrpqVLl+rVV19Vp06d9PjjjysnJ0e33367v7sCAAAhyO/rfEjSwIEDNXDgwECcGgAAhDh+2wUAAFhF+AAAAFYRPgAAgFUBuebjXMEiWAAA1B4zHwAAwCrCBwAAsIrwAQAArCJ8AAAAqwgfAADAKu52OQMHxmYFu4R6hzuIAKD+Y+YDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYFVEsAsIZY2mTw12CQAAhBxmPgAAgFWEDwAAYBXhAwAAWEX4AAAAVnHB6Rk4MDYr2CX4DRfPAgBsYeYDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYFPHxkZ2fL5XJp9OjRge4KAACEgICGj3Xr1mnevHm65JJLAtkNAAAIIQELH4cOHdLtt9+u+fPnq1GjRoHqBgAAhJiAhY8RI0ZowIABuvrqq2tsV1JSoqKiIq8HAACovwKyvPprr72m/Px8rVu37pRts7OzNWnSpECUAQAAzkJ+n/nYtWuXRo0apYULFyo6OvqU7cePH6/CwkLnsWvXLn+XBAAAziJ+n/nIy8tTQUGBunbt6mwrLy/X6tWrNXv2bJWUlCg8PNzZ53a75Xa7/V0GAAA4S/k9fPzqV7/SF1984bXtrrvuUvv27TVu3Div4AEAAM49fg8fcXFx6tSpk9e2Bg0aKDk5ucp2AABw7mGFUwAAYFVA7nY52cqVK210AwAAQgAzHwAAwCrCBwAAsIrwAQAArLJyzUd91Wj61GCXAABAyGHmAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYxd0uZ+DA2Kxgl+A33LkDALCFmQ8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFeEDAABYRfgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVRHBLiCUNZo+NdglAAAQcpj5AAAAVhE+AACAVYQPAABgFeEDAABYxQWnZ+DA2Kxgl+A3XDwLALCFmQ8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWOX38JGdna1u3bopLi5OTZo00Y033qhvv/3W390AAIAQ5ffwsWrVKo0YMUKffPKJcnNzdfz4cfXv31+HDx/2d1cAACAE+X2RsX/84x9ezxcsWKAmTZooLy9PvXv3rtK+pKREJSUlzvOioiJ/lwQAgBUVFRUqLS0NdhkBExkZqfDw8DM+T8BXOC0sLJQkJSUl+dyfnZ2tSZMmBboMAAACqrS0VNu3b1dFRUWwSwmoxMREpaamyuVy1fkcLmOM8WNNXowxuuGGG3TgwAGtWbPGZxtfMx9paWkqLCxUfHx8oErzC5ZXBwBIv3ze7dy5U2VlZfJ4PAoLq3/3cxhjdOTIERUUFCgxMVFNmzb12l9UVKSEhITT+vwO6MzHAw88oM8//1xr166tto3b7Zbb7Q5kGQAABNTx48d15MgReTwexcbGBrucgImJiZEkFRQUqEmTJnX+CiZg0ezBBx/U8uXLtWLFCjVv3jxQ3QAAEHTl5eWSpKioqCBXEniV4aqsrKzO5/D7zIcxRg8++KCWLl2qlStXqnXr1v7uAgCAs9KZXAcRKvwxRr+HjxEjRmjRokV68803FRcXp71790qSEhISnOkaAABw7vJ7+Jg7d64kKSMjw2v7ggULdOedd/q7OwAAzlrHd+9Wxf791voLS0pSRLNm1vqrq4B87QIAwLnu+O7d+vGqPtIJd3QGnNutlDWrah1A5syZoyeffFJ79uxRx44dlZOTo6uuuipARfLbLgAABETF/v12g4cklZTUeqZl8eLFGj16tCZMmKANGzboqquuUmZmpnbu3BmgIgkfAACc02bMmKHf//73uueee9ShQwfl5OQoLS3NuYwiEAgfAACco0pLS5WXl6f+/ft7be/fv78+/vjjgPUb8OXV6zNWBQUAhLKff/5Z5eXlSklJ8dqekpLi3K0aCMx8AABwjjt57Q5jTEDXLCF8AABwjmrcuLHCw8OrzHIUFBRUmQ3xJ8IHAADnqKioKHXt2lW5uble23Nzc9WzZ8+A9cs1HwAAnMPGjBmjIUOGKD09XVdccYXmzZunnTt3avjw4QHrk/ABAKcwdfnmYJdwVsm6vmOwS4Af3XLLLdq3b58mT56sPXv2qFOnTnrnnXfUsmXLgPVJ+AAAIADCkpIkt9v6CqdhSUm1Puz+++/X/fffH4CCfCN8AAAQABHNmillzSp+28UHwgcAAAES0ayZFAJhwDbudgEAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVYQPAABgFet8AAAQIHsPHtXBI2XW+kuMjVRqYoy1/uqK8AEAQADsPXhUv5u1VqXHK6z1GRURpr8+eGWtAsjq1av15JNPKi8vT3v27NHSpUt14403Bq5I8bULAAABcfBImdXgIUmlxytqPdNy+PBhde7cWbNnzw5QVVUx8wEAwDksMzNTmZmZVvtk5gMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWMXdLgAAnMMOHTqkbdu2Oc+3b9+ujRs3KikpSS1atAhIn4QPADiFrOs7BrsEIGDWr1+vvn37Os/HjBkjSRo2bJhefPHFgPRJ+AAAIAASYyMVFRFmfYXTxNjIWh2TkZEhY0yAKvKN8AEAQACkJsborw9eyW+7+ED4AAAgQFITY0IiDNjG3S4AAMAqwgcAALCKr10AANZNXb452CX4VWxEhbqnGBUUHVPkscBcvNm0Hn19w8wHAAB+YSTLd40EQ0XFmd+9w8wHAABnqKTcpdLjRoeKDqhhfCPJ5fJ7H8eO+f+ctWGMUWlpqX766SeFhYUpKiqqzucifAAAcIbKjUub94ero4pUVFQsyf9B4XAt1+8IlNjYWLVo0UJhYXX/8oTwAQCAHxSWhenTApfc4YH56uW/+7UOyHlrIzw8XBEREXKd4cwO4QMAAD8pNy4dOR6Yr0eio6MDct5gCNgFp3PmzFHr1q0VHR2trl27as2aNYHqCgAAhJCAhI/Fixdr9OjRmjBhgjZs2KCrrrpKmZmZ2rlzZyC6AwAAISQg4WPGjBn6/e9/r3vuuUcdOnRQTk6O0tLSNHfu3EB0BwAAQojfr/koLS1VXl6esrKyvLb3799fH3/8cZX2JSUlKikpcZ4XFhZKkoqKivxdGgDgLHHsyKFglxByzvbPxcr6TucXcv0ePn7++WeVl5crJSXFa3tKSor27t1bpX12drYmTZpUZXtaWpq/SwMAIGRV/aQ8OxUXFyshIaHGNgG72+Xk23CMMT5vzRk/frzGjBnjPK+oqND+/fuVnJx8xrfyBFJRUZHS0tK0a9cuxcfHB7ucgGCM9QNjrB8YY/1Qn8dojFFxcbE8Hs8p2/o9fDRu3Fjh4eFVZjkKCgqqzIZIktvtltvt9tqWmJjo77ICJj4+vt79AzoZY6wfGGP9wBjrh/o6xlPNeFTy+wWnUVFR6tq1q3Jzc7225+bmqmfPnv7uDgAAhJiAfO0yZswYDRkyROnp6briiis0b9487dy5U8OHDw9EdwAAIIQEJHzccsst2rdvnyZPnqw9e/aoU6dOeuedd9SyZctAdBcUbrdbjz32WJWvjOoTxlg/MMb6gTHWD+fCGE+Hy5zOPTEAAAB+ErDl1QEAAHwhfAAAAKsIHwAAwCrCBwAAsIrwAQAArCJ81MGcOXPUunVrRUdHq2vXrlqzZk2wS6qz7OxsdevWTXFxcWrSpIluvPFGffvtt15tjDGaOHGiPB6PYmJilJGRoc2bNwep4jOXnZ0tl8ul0aNHO9vqwxh3796tO+64Q8nJyYqNjVWXLl2Ul5fn7A/1MR4/flx//OMf1bp1a8XExKhNmzaaPHmyKioqnDahNsbVq1dr0KBB8ng8crlcWrZsmdf+0xlPSUmJHnzwQTVu3FgNGjTQ9ddfr++//97iKGpW0xjLyso0btw4XXzxxWrQoIE8Ho+GDh2qH374wescoTzGk917771yuVzKycnx2n62j9HfCB+1tHjxYo0ePVoTJkzQhg0bdNVVVykzM1M7d+4Mdml1smrVKo0YMUKffPKJcnNzdfz4cfXv31+HDx922kyfPl0zZszQ7NmztW7dOqWmpuqaa65RcXFxECuvm3Xr1mnevHm65JJLvLaH+hgPHDigXr16KTIyUu+++66++uorPf30014/VRDqY5w2bZqef/55zZ49W19//bWmT5+uJ598UrNmzXLahNoYDx8+rM6dO2v27Nk+95/OeEaPHq2lS5fqtdde09q1a3Xo0CENHDhQ5eXltoZRo5rGeOTIEeXn5+tPf/qT8vPztWTJEm3ZskXXX3+9V7tQHuOJli1bpk8//dTnb5+c7WP0O4Na6d69uxk+fLjXtvbt25usrKwgVeRfBQUFRpJZtWqVMcaYiooKk5qaaqZOneq0OXbsmElISDDPP/98sMqsk+LiYtOuXTuTm5tr+vTpY0aNGmWMqR9jHDdunLnyyiur3V8fxjhgwABz9913e2276aabzB133GGMCf0xSjJLly51np/OeA4ePGgiIyPNa6+95rTZvXu3CQsLM//4xz+s1X66Th6jL5999pmRZHbs2GGMqT9j/P77702zZs3Ml19+aVq2bGmeeeYZZ1+ojdEfmPmohdLSUuXl5al///5e2/v376+PP/44SFX5V2FhoSQpKSlJkrR9+3bt3bvXa8xut1t9+vQJuTGPGDFCAwYM0NVXX+21vT6Mcfny5UpPT9fNN9+sJk2a6NJLL9X8+fOd/fVhjFdeeaU+/PBDbdmyRZK0adMmrV27Vtddd52k+jHGE53OePLy8lRWVubVxuPxqFOnTiE5ZumXv0Eul8uZtasPY6yoqNCQIUP0yCOPqGPHjlX214cx1lZAllevr37++WeVl5dX+XXelJSUKr/iG4qMMRozZoyuvPJKderUSZKccfka844dO6zXWFevvfaa8vPztW7duir76sMY//Of/2ju3LkaM2aM/vCHP+izzz7TyJEj5Xa7NXTo0HoxxnHjxqmwsFDt27dXeHi4ysvL9cQTT+jWW2+VVD/exxOdznj27t2rqKgoNWrUqEqbUPybdOzYMWVlZem2225zfvG1Poxx2rRpioiI0MiRI33urw9jrC3CRx24XC6v58aYKttC0QMPPKDPP/9ca9eurbIvlMe8a9cujRo1Su+//76io6OrbRfKY6yoqFB6erqmTJkiSbr00ku1efNmzZ07V0OHDnXahfIYFy9erIULF2rRokXq2LGjNm7cqNGjR8vj8WjYsGFOu1Aeoy91GU8ojrmsrEyDBw9WRUWF5syZc8r2oTLGvLw8Pfvss8rPz691vaEyxrrga5daaNy4scLDw6sk0YKCgir/dRJqHnzwQS1fvlwrVqxQ8+bNne2pqamSFNJjzsvLU0FBgbp27aqIiAhFRERo1apVmjlzpiIiIpxxhPIYmzZtqosuushrW4cOHZwLoevD+/jII48oKytLgwcP1sUXX6whQ4booYceUnZ2tqT6McYTnc54UlNTVVpaqgMHDlTbJhSUlZXpd7/7nbZv367c3Fxn1kMK/TGuWbNGBQUFatGihfP3Z8eOHXr44YfVqlUrSaE/xrogfNRCVFSUunbtqtzcXK/tubm56tmzZ5CqOjPGGD3wwANasmSJPvroI7Vu3dprf+vWrZWamuo15tLSUq1atSpkxvyrX/1KX3zxhTZu3Og80tPTdfvtt2vjxo1q06ZNyI+xV69eVW6R3rJli/NL0vXhfTxy5IjCwrz/ZIWHhzu32taHMZ7odMbTtWtXRUZGerXZs2ePvvzyy5AZc2Xw2Lp1qz744AMlJyd77Q/1MQ4ZMkSff/65198fj8ejRx55RO+9956k0B9jnQTpQteQ9dprr5nIyEjzwgsvmK+++sqMHj3aNGjQwHz33XfBLq1O7rvvPpOQkGBWrlxp9uzZ4zyOHDnitJk6dapJSEgwS5YsMV988YW59dZbTdOmTU1RUVEQKz8zJ97tYkzoj/Gzzz4zERER5oknnjBbt241r7zyiomNjTULFy502oT6GIcNG2aaNWtm3n77bbN9+3azZMkS07hxYzN27FinTaiNsbi42GzYsMFs2LDBSDIzZswwGzZscO70OJ3xDB8+3DRv3tx88MEHJj8/3/Tr18907tzZHD9+PFjD8lLTGMvKysz1119vmjdvbjZu3Oj1N6ikpMQ5RyiP0ZeT73Yx5uwfo78RPurgueeeMy1btjRRUVHmsssuc25LDUWSfD4WLFjgtKmoqDCPPfaYSU1NNW632/Tu3dt88cUXwSvaD04OH/VhjG+99Zbp1KmTcbvdpn379mbevHle+0N9jEVFRWbUqFGmRYsWJjo62rRp08ZMmDDB60Mq1Ma4YsUKn///GzZsmDHm9MZz9OhR88ADD5ikpCQTExNjBg4caHbu3BmE0fhW0xi3b99e7d+gFStWOOcI5TH64it8nO1j9DeXMcbYmGEBAACQuOYDAABYRvgAAABWET4AAIBVhA8AAGAV4QMAAFhF+AAAAFYRPgAAgFWEDwAAYBXhAwAAWEX4AAAAVhE+AACAVf8feyO0PeznshgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Adjusting the size of the plot\n",
    "plt.figure(figsize=(1, 1))  # You can adjust these numbers as per your needs\n",
    "# Adjusting the layout\n",
    "plt.tight_layout()\n",
    "gd.plot_persistence_barcode(persistence3, legend=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a5691869-1d65-4932-b56b-80534e1494e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<networkx.classes.graph.Graph at 0x794c32ef6650>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visualize_simplicial_complex(simplex_tree3,100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TDA",
   "language": "python",
   "name": "tda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
